<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Phylo - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Phylo";
    var mkdocs_page_input_path = "phylo.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_assemble/">Assemble</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Description</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Phylo</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#phylo-quick-start">Phylo Quick-Start</a></li>
    

    <li class="toctree-l2"><a href="#multiple_align">multiple_align</a></li>
    

    <li class="toctree-l2"><a href="#model_test">model_test</a></li>
    

    <li class="toctree-l2"><a href="#build_tree_ng">build_tree_NG</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../inout/">File Input/Output</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Phylo</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>This stage includes phylogenomics stages.</p>
<h3 id="phylo-quick-start">Phylo Quick-Start</h3>
<p>HAPHPIPE includes three modules for phylogenomics: <em>multiple_align</em>, <em>model_test</em>, and <em>build_tree</em>. These three modules are sufficient to turn your consensus and/or haplotype sequences from the other modules into a phylogenetic tree! For purposes of this quick-start guide, we will demonstrate the modules to create a tree from HIV <em>pol</em> consensus sequences. </p>
<p><strong>Step 1: Alignment</strong></p>
<p>After running either of the assembly pipelines, <code>final.fna</code> files will be located in directories named <code>./&lt;SampleID&gt;/haphpipe_assemble_0[1|2]</code>. For the <em>multiple_align</em> module, we need to create a list of all of these directories. We can do so easily with one command (shown for <code>haphpipe_assemble_01</code> output:</p>
<pre><code>ls -d ./SRR*/haphpipe_assemble_01 &gt; ./dir_list.txt
</code></pre>

<p>Now, we will align all of these <code>final.fna</code> files:</p>
<pre><code>haphpipe multiple_align --dir_list dir_list.txt --ref_gtf refs/HIV_B.K03455.HXB2.gtf 
</code></pre>

<p>The output will be located in a new directory, <code>hp_multiple_align</code>. The alignment of <em>pol</em> sequences is the file <code>alignment_region00.fasta</code>.</p>
<p><strong>Step 2: Model Selection</strong></p>
<p>Now, we will use the <em>model_test</em> module to determine the best-fit evolutionary model for our data. This is an input to the tree building module. We will use this command to generate best-fit models available in RAxML:</p>
<pre><code>haphpipe model_test --seqs hp_multiple_align/alignment_region00.fasta --run_id alignment_region00 --template raxml
</code></pre>

<p>The ModelTest output will be written to a file called <code>modeltest_results.out</code> and a summary of all the best models will be written to <code>modeltest_results_summary.tsv</code>. Examples of both are below.</p>
<details>
  <summary>ModelTest-NG Output</summary>

<pre><code>    --------------------------------------------------------------------------------
    ModelTest-NG vx.y.z

    Input data:
      MSA:        multiple_align/alignment.fasta
      Tree:       Maximum likelihood
        file:           -
      #taxa:            6
      #sites:           1975
      #patterns:        114
      Max. thread mem:  0 MB

    Output:
      Log:           /var/folders/lv/dqdkd8957_3fv6yxsyfsvn0r0000gn/T/tmpHP_model_testud4sc4ya/samp12_modeltest_results.log
      Starting tree: /var/folders/lv/dqdkd8957_3fv6yxsyfsvn0r0000gn/T/tmpHP_model_testud4sc4ya/samp12_modeltest_results.tree
      Results:       /var/folders/lv/dqdkd8957_3fv6yxsyfsvn0r0000gn/T/tmpHP_model_testud4sc4ya/samp12_modeltest_results.out

    Selection options:
      # dna schemes:      11
      # dna models:       88
      include model parameters:
        Uniform:         true
        p-inv (+I):      true
        gamma (+G):      true
        both (+I+G):     true
        free rates (+R): false
        fixed freqs:     true
        estimated freqs: true
        #categories:     4
      gamma rates mode:   mean
      asc bias:           none
      epsilon (opt):      0.01
      epsilon (par):      0.05
      keep branches:      false

    Additional options:
      verbosity:        very low
      threads:          1/6
      RNG seed:         12345
      subtree repeats:  enabled
    --------------------------------------------------------------------------------

    BIC       model              K            lnL          score          delta    weight
    --------------------------------------------------------------------------------
           1  HKY                4     -5380.7535     10860.1551         0.0000    0.7049
           2  TrN                5     -5378.2013     10862.6391         2.4839    0.2036
           3  TPM1uf             5     -5379.8699     10865.9763         5.8211    0.0384
           4  TPM3uf             5     -5380.7175     10867.6716         7.5164    0.0164
           5  HKY+G4             5     -5380.8440     10867.9246         7.7694    0.0145
           6  HKY+I              5     -5381.4917     10869.2200         9.0648    0.0076
           7  TIM3               6     -5378.1637     10870.1523         9.9971    0.0048
           8  TrN+G4             6     -5378.3069     10870.4387        10.2836    0.0041
           9  TPM2uf+G4          6     -5378.9908     10871.8064        11.6512    0.0021
          10  TrN+I              6     -5379.0181     10871.8610        11.7059    0.0020
    --------------------------------------------------------------------------------
    Best model according to BIC
    ---------------------------
    Model:              HKY
    lnL:                -5380.7535
    Frequencies:        0.3695 0.1709 0.2219 0.2377
    Subst. Rates:       1.0000 2.4741 1.0000 1.0000 2.4741 1.0000 
    Inv. sites prop:    -
    Gamma shape:        -
    Score:              10860.1551
    Weight:             0.7049
    ---------------------------
    Parameter importances
    ---------------------------
    P.Inv:              0.0100
    Gamma:              0.0216
    Gamma-Inv:          0.0002
    Frequencies:        1.0000
    ---------------------------
    Model averaged estimates
    ---------------------------
    P.Inv:              0.0215
    Alpha:              94.2337
    Alpha-P.Inv:        91.7960
    P.Inv-Alpha:        0.0214
    Frequencies:        0.3700 0.1702 0.2226 0.2372 

    Commands:
      &gt; phyml  -i multiple_align/alignment.fasta -m 010010 -f m -v 0 -a 0 -c 1 -o tlr
      &gt; raxmlHPC-SSE3 -s multiple_align/alignment.fasta -c 1 -m GTRCATX -n EXEC_NAME -p PARSIMONY_SEED
      &gt; raxml-ng --msa multiple_align/alignment.fasta --model HKY
      &gt; paup -s multiple_align/alignment.fasta
      &gt; iqtree -s multiple_align/alignment.fasta -m HKY

    AIC       model              K            lnL          score          delta    weight
    --------------------------------------------------------------------------------
           1  TrN                5     -5378.2013     10784.4026         0.0000    0.2246
           2  TIM2+G4            7     -5376.4972     10784.9944         0.5919    0.1671
           3  TIM3               6     -5378.1637     10786.3274         1.9249    0.0858
           4  TIM2+I             7     -5377.2570     10786.5141         2.1115    0.0782
           5  TrN+G4             6     -5378.3069     10786.6138         2.2113    0.0744
           6  TIM1+G4            7     -5377.3549     10786.7098         2.3073    0.0709
           7  HKY                4     -5380.7535     10787.5069         3.1044    0.0476
           8  TPM1uf             5     -5379.8699     10787.7398         3.3372    0.0423
           9  TPM2uf+G4          6     -5378.9908     10787.9815         3.5790    0.0375
          10  TrN+I              6     -5379.0181     10788.0362         3.6336    0.0365
    --------------------------------------------------------------------------------
    Best model according to AIC
    ---------------------------
    Model:              TrN
    lnL:                -5378.2013
    Frequencies:        0.3720 0.1679 0.2249 0.2352
    Subst. Rates:       1.0000 2.2008 1.0000 1.0000 3.1065 1.0000 
    Inv. sites prop:    -
    Gamma shape:        -
    Score:              10784.4026
    Weight:             0.2246
    ---------------------------
    Parameter importances
    ---------------------------
    P.Inv:              0.1262
    Gamma:              0.3943
    Gamma-Inv:          0.0384
    Frequencies:        1.0000
    ---------------------------
    Model averaged estimates
    ---------------------------
    P.Inv:              0.0216
    Alpha:              93.2595
    Alpha-P.Inv:        94.4193
    P.Inv-Alpha:        0.0216
    Frequencies:        0.3718 0.1684 0.2238 0.2359 

    Commands:
      &gt; phyml  -i multiple_align/alignment.fasta -m 010020 -f m -v 0 -a 0 -c 1 -o tlr
      &gt; raxmlHPC-SSE3 -s multiple_align/alignment.fasta -c 1 -m GTRCATX -n EXEC_NAME -p PARSIMONY_SEED
      &gt; raxml-ng --msa multiple_align/alignment.fasta --model TrN
      &gt; paup -s multiple_align/alignment.fasta
      &gt; iqtree -s multiple_align/alignment.fasta -m TrN

    AICc      model              K            lnL          score          delta    weight
    --------------------------------------------------------------------------------
           1  TrN                5     -5378.2013     10784.4026         0.0000    0.2246
           2  TIM2+G4            7     -5376.4972     10784.9944         0.5919    0.1671
           3  TIM3               6     -5378.1637     10786.3274         1.9249    0.0858
           4  TIM2+I             7     -5377.2570     10786.5141         2.1115    0.0782
           5  TrN+G4             6     -5378.3069     10786.6138         2.2113    0.0744
           6  TIM1+G4            7     -5377.3549     10786.7098         2.3073    0.0709
           7  HKY                4     -5380.7535     10787.5069         3.1044    0.0476
           8  TPM1uf             5     -5379.8699     10787.7398         3.3372    0.0423
           9  TPM2uf+G4          6     -5378.9908     10787.9815         3.5790    0.0375
          10  TrN+I              6     -5379.0181     10788.0362         3.6336    0.0365
    --------------------------------------------------------------------------------
    Best model according to AICc
    ---------------------------
    Model:              TrN
    lnL:                -5378.2013
    Frequencies:        0.3720 0.1679 0.2249 0.2352
    Subst. Rates:       1.0000 2.2008 1.0000 1.0000 3.1065 1.0000 
    Inv. sites prop:    -
    Gamma shape:        -
    Score:              10784.4026
    Weight:             0.2246
    ---------------------------
    Parameter importances
    ---------------------------
    P.Inv:              0.1262
    Gamma:              0.3943
    Gamma-Inv:          0.0384
    Frequencies:        1.0000
    ---------------------------
    Model averaged estimates
    ---------------------------
    P.Inv:              0.0216
    Alpha:              93.2595
    Alpha-P.Inv:        94.4193
    P.Inv-Alpha:        0.0216
    Frequencies:        0.3718 0.1684 0.2238 0.2359 

    Commands:
      &gt; phyml  -i multiple_align/alignment.fasta -m 010020 -f m -v 0 -a 0 -c 1 -o tlr
      &gt; raxmlHPC-SSE3 -s multiple_align/alignment.fasta -c 1 -m GTRCATX -n EXEC_NAME -p PARSIMONY_SEED
      &gt; raxml-ng --msa multiple_align/alignment.fasta --model TrN
      &gt; paup -s multiple_align/alignment.fasta
      &gt; iqtree -s multiple_align/alignment.fasta -m TrN
    Done

</code></pre>

</details>

<details>
  <summary>ModelTest-NG Output Summary</summary>

<pre><code>    File    Criteria    Best Model
    hp_multiple_align/alignment.fasta   BIC HKY
    hp_multiple_align/alignment.fasta   AIC TrN
    hp_multiple_align/alignment.fasta   AICc    TrN
</code></pre>

</details>

<p><br></p>
<p><strong>Step 3: Build a Tree</strong></p>
<p>Now, we will use <em>build_tree_NG_</em> to build our tree! You should use the best model outputted in <em>model_test</em> for the <code>--model</code> argument (here we are using GTR). The <code>--all</code> option will automatically run a full maximum likelihood &amp; bootstrapping analysis for us:</p>
<pre><code>haphpipe build_tree_NG --seqs hp_multiple_align/alignment_region00.fasta --all --model GTR
</code></pre>

<p>The output will be written to a new directory, <code>hp_build_tree</code>. The best tree file from RAxML will be outputted as <code>hp_tree.raxml.support</code>. This tree can then be annotated in programs such as <a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> or <a href="https://itol.embl.de">iTOL</a>.</p>
<p><strong>Phylogenomics Pipelines</strong></p>
<p>For users who would like to build a full pipeline to run assembly and phylogenetics stages in one go, we recommend adapting the demo pipeline (<code>haphpipe_demo</code>) for this purpose. See the <a href="https://gwcbi.github.io/haphpipe_docs/demos/">demo page</a> for more details.</p>
<h3 id="multiple_align"><em>multiple_align</em></h3>
<p>Align consensus sequences using MAFFT (<a href="https://mafft.cbrc.jp/alignment/software/manual/manual.html">documentation</a>). Input can be a list of directories which contain <code>final.fna</code> and/or <code>ph_haplotypes.fna</code> files or a fasta file, or both (in which case the sequences in the FASTA file are combined with the <code>final.fna</code> and/or <code>ph_haplotypes.fna</code> files retreived before the alignment.
Sequences will be separated by amplicons using a supplied GTF file before alignment (unless the <code>--alignall</code> option is specified). This module may also be used to separate files by amplicons (without aligning) by specifying the <code>--fastaonly</code> option.
Alignments are by default outputted as FASTA files, although PHYLIP (<code>--phylipout</code>) or CLUSTAL (<code>--clustalout</code>) output options are also available.
Many options from MAFFT are available in this module. Please refer to the MAFFT documentation above for information about these options.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe multiple_align [MAFFT OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --dir_list &lt;TXT&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_multiple_align [MAFFT OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --dir_list &lt;TXT&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><em>Output files:</em> alignment files in FASTA format (default), one per amplicon (or one <code>alignment.fasta</code> file if using <code>--alignall</code> option)</p>
<p>Note: MAFFT stores intermediate files in a temporary directory located in /tmp. More information is available <a href="https://mafft.cbrc.jp/alignment/software/mpi.html#TMPDIR">here</a>.</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqs SEQS</td>
<td>FASTA file with sequences to be aligned</td>
</tr>
<tr>
<td>--dir_list DIR_LIST</td>
<td>List of directories which include either a final.fna or ph_haplotypes.fna file, one on each line</td>
</tr>
<tr>
<td>--ref_gtf REF_GTF</td>
<td>Reference GTF file</td>
</tr>
<tr>
<td>--out_align OUT_ALIGN</td>
<td>Name for alignment file</td>
</tr>
<tr>
<td>--nuc</td>
<td>Assume nucleotide (default: False)</td>
</tr>
<tr>
<td>--amino</td>
<td>Assume amino (default: False)</td>
</tr>
<tr>
<td>--clustalout</td>
<td>Clustal output format (default: False)</td>
</tr>
<tr>
<td>--phylipout</td>
<td>PHYLIP output format (default: False)</td>
</tr>
<tr>
<td>--inputorder</td>
<td>Output order same as input (default: False)</td>
</tr>
<tr>
<td>--reorder</td>
<td>Output order aligned (default: False)</td>
</tr>
<tr>
<td>--treeout</td>
<td>Guide tree is output to the input.tree file (default:False)</td>
</tr>
<tr>
<td>--quiet_mafft</td>
<td>Do not report progress (default: False)</td>
</tr>
<tr>
<td>--outdir OUTDIR</td>
<td>Output directory</td>
</tr>
</tbody>
</table>
<p><em>MAFFT Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--algo ALGO</td>
<td>Use different algorithm in command: linsi, ginsi,   einsi, fftnsi, fftns, nwns, nwnsi</td>
</tr>
<tr>
<td>--auto</td>
<td>Automatically select algorithm (default: False)</td>
</tr>
<tr>
<td>--sixmerpair</td>
<td>Calculate distance based on shared 6mers, on by default (default: False)</td>
</tr>
<tr>
<td>--globalpair</td>
<td>Use Needleman-Wunsch algorithm (default: False)</td>
</tr>
<tr>
<td>--localpair</td>
<td>Use Smith-Waterman algorithm (default: False)</td>
</tr>
<tr>
<td>--genafpair</td>
<td>Use local algorithm with generalized affine gap cost (default: False)</td>
</tr>
<tr>
<td>--fastapair</td>
<td>Use FASTA for pairwise alignment (default: False)</td>
</tr>
<tr>
<td>--weighti WEIGHTI</td>
<td>Weighting factor for consistency term</td>
</tr>
<tr>
<td>--retree RETREE</td>
<td>Number of times to build guide tree</td>
</tr>
<tr>
<td>--maxiterate MAXITERATE</td>
<td>Number of cycles for iterative refinement</td>
</tr>
<tr>
<td>--noscore</td>
<td>Do not check alignment score in iterative alignment (default: False)</td>
</tr>
<tr>
<td>--memsave</td>
<td>Use Myers-Miller algorithm (default: False)</td>
</tr>
<tr>
<td>--parttree</td>
<td>Use fast tree-building method with 6mer distance (default: False)</td>
</tr>
<tr>
<td>--dpparttree</td>
<td>Use PartTree algorithm with distances based on DP (default: False)</td>
</tr>
<tr>
<td>--fastaparttree</td>
<td>Use PartTree algorithm with distances based on FASTA (default: False)</td>
</tr>
<tr>
<td>--partsize PARTSIZE</td>
<td>Number of partitions for PartTree</td>
</tr>
<tr>
<td>--groupsize GROUPSIZE</td>
<td>Max number of sequences for PartTree</td>
</tr>
</tbody>
</table>
<p><em>MAFFT Parameters:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--lop LOP</td>
<td>Gap opening penalty</td>
</tr>
<tr>
<td>--lep LEP</td>
<td>Offset value</td>
</tr>
<tr>
<td>--lexp LEXP</td>
<td>Gap extension penalty</td>
</tr>
<tr>
<td>--LOP LOP</td>
<td>Gap opening penalty to skip alignment</td>
</tr>
<tr>
<td>--LEXP LEXP</td>
<td>Gap extension penalty to skip alignment</td>
</tr>
<tr>
<td>--bl BL</td>
<td>BLOSUM matrix: 30, 45, 62, or 80</td>
</tr>
<tr>
<td>--jtt JTT</td>
<td>JTT PAM number &gt;0</td>
</tr>
<tr>
<td>--tm TM</td>
<td>Transmembrane PAM number &gt;0</td>
</tr>
<tr>
<td>--aamatrix AAMATRIX</td>
<td>Path to user-defined AA scoring matrix</td>
</tr>
<tr>
<td>--fmodel</td>
<td>Incorporate AA/nuc composition info into scoring matrix (default: False)</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu NCPU</td>
<td>Number of CPU to use (default: 1)</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and  stderr) (default: False)</td>
</tr>
<tr>
<td>--logfile LOGFILE</td>
<td>Name for log file (output)</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False)</td>
</tr>
<tr>
<td>--fastaonly</td>
<td>Output fasta files separated by region but do not     align (default: False)</td>
</tr>
<tr>
<td>--alignall</td>
<td>Do not separate files by region, align entire file (default: False)</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe multiple_align --dir_list demo_sra_list.txt --ref_gtf HIV_B.K03455.HXB2.gtf --phylipout --logfile demo_multiple_align.log
</code></pre>

<h3 id="model_test"><em>model_test</em></h3>
<p>Select the best-fit model of evolution from an alignment file using ModelTest-NG (<a href="https://github.com/ddarriba/modeltest/wiki">documentation</a>). Input is an alignment in FASTA or PHYLIP format. Output is ModelTest-NG results (text file) containing information for the best performing models.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe model_test [ModelTest-NG OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_model_test [ModelTest-NG OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> ModelTest-NG output file (<code>modeltest_results.out</code>).</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqs SEQS</td>
<td>Alignment in FASTA or PHYLIP format</td>
</tr>
<tr>
<td>--outname</td>
<td>Name for output file</td>
</tr>
<tr>
<td>--outdir OUTDIR</td>
<td>Output directory</td>
</tr>
</tbody>
</table>
<p><em>ModelTest-NG Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--data_type</td>
<td>Data type: nt or aa (default: nt)</td>
</tr>
<tr>
<td>--partitions</td>
<td>Partitions file</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator</td>
</tr>
<tr>
<td>--topology TOPOLOGY</td>
<td>Starting topology: ml, mp, fixed-ml-jc, fixed-ml-gtr, fixed-mp, random, or user (default: ml)</td>
</tr>
<tr>
<td>--utree</td>
<td>User-defined starting tree</td>
</tr>
<tr>
<td>--force</td>
<td>Force output overriding (default: False)</td>
</tr>
<tr>
<td>--asc_bias ASC_BIAS</td>
<td>Ascertainment bias correction: lewis, felsenstein, or stamatakis</td>
</tr>
<tr>
<td>--frequencies FREQUENCIES</td>
<td>Candidate model frequencies: e (estimated) or f (fixed)</td>
</tr>
<tr>
<td>--het HET</td>
<td>Set rate heterogeneity: u (uniform), i (invariant sites +I), g (gamma +G), or f (bothinvariant sites and gamma +I+G)</td>
</tr>
<tr>
<td>--models MODELS</td>
<td>Text file with candidate models, one per line</td>
</tr>
<tr>
<td>--schemes SCHEMES</td>
<td>Number of predefined DNA substitution schemes evaluated: 3, 5, 7, 11, or 203</td>
</tr>
<tr>
<td>--template TEMPLATE</td>
<td>Set candidate models according to a specified tool: raxml, phyml, mrbayes, or paup</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu NCPU</td>
<td>Number of CPU to use (default: 1)</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and  stderr) (default: False)</td>
</tr>
<tr>
<td>--logfile LOGFILE</td>
<td>Name for log file (output)</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False)</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False)</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe model_test --seqs multiple_align/alignment.fasta
</code></pre>

<h3 id="build_tree_ng"><em>build_tree_NG</em></h3>
<p>Phylogeny reconstruction with RAxML-NG (<a href="https://github.com/amkozlov/raxml-ng/wiki">documentation</a>). Input is an alignment (FASTA or PHYLIP format). Output is a tree file.
Please see the RAxML-NG documentation for a full description of RAxML-NG options. </p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe build_tree_NG [RAxML OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --output_name &lt;TXT&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp build_tree_NG [RAxML OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --output_name &lt;TXT&gt; [--outdir]</code></p>
<p><em>Output files:</em></p>
<table>
<thead>
<tr>
<th>File</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>%PREFIX.raxml.bestTree</td>
<td>Outputs the best-scoring Maximum Likelihood tree</td>
</tr>
<tr>
<td>%PREFIX.raxml.bestPartitionTrees</td>
<td>Best-scoring ML tree for each partition</td>
</tr>
<tr>
<td>%PREFIX.raxml.bestModel</td>
<td>Optimized parameters for the highest-scoring ML tree</td>
</tr>
<tr>
<td>%PREFIX.raxml.bootstraps</td>
<td>Trees generated for every bootstrap replicate</td>
</tr>
<tr>
<td>%PREFIX.raxml.bootstrapMSA.<REP>.phy</td>
<td>Bootstrap replicate alignments</td>
</tr>
<tr>
<td>%PREFIX.raxml.ckp</td>
<td>Contains last log record if RAxML-NG has not finished successfully</td>
</tr>
<tr>
<td>%PREFIX.raxml.consensusTree</td>
<td>Consensus tree: estimates support for each clade of the final tree</td>
</tr>
<tr>
<td>%PREFIX.raxml.log</td>
<td>Screen log</td>
</tr>
<tr>
<td>%PREFIX.raxml.mlTrees</td>
<td>Maximum Likelihood trees for each starting tree</td>
</tr>
<tr>
<td>%PREFIX.raxml.startTree</td>
<td>The starting trees for each Maximum Likelihood inference</td>
</tr>
<tr>
<td>%PREFIX.raxml.support</td>
<td>Best-scoring Maximum Likelihood tree with bootstrap values</td>
</tr>
<tr>
<td>%PREFIX.raxml.terrace</td>
<td>Trees residing on a terrace (same likelihood or parsimony score) in compressed Newick form</td>
</tr>
<tr>
<td>%PREFIX.raxml.terraceNewick</td>
<td>Trees residing on a terrace in multi-line standard Newick form</td>
</tr>
</tbody>
</table>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>RAxML-NG Equivalent</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqs SEQS</td>
<td>--msa</td>
<td>Input alignment in PHYLIP or FASTA format</td>
</tr>
<tr>
<td>--output_name NAME</td>
<td>--prefix</td>
<td>Run name for trees (default: build_tree.tre)</td>
</tr>
<tr>
<td>--outdir OUTDIR</td>
<td>--prefix</td>
<td>Output directory (default: .)</td>
</tr>
</tbody>
</table>
<p><em>RAxML-NG Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>RAxML Equivalent-NG</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--model MODEL</td>
<td>--model</td>
<td>Substitution model OR path to partition file</td>
</tr>
<tr>
<td>--all</td>
<td>--all</td>
<td>Run bootstrap search and find best ML tree</td>
</tr>
<tr>
<td>--branch_length BRANCH_LENGTH</td>
<td>--brlen</td>
<td>Specify branch linkage model</td>
</tr>
<tr>
<td>--consense</td>
<td>--consense</td>
<td>Build a consensus tree</td>
</tr>
<tr>
<td>--rand_tree RAND_TREE</td>
<td>--tree rand</td>
<td>Start tree option: start from a random topology</td>
</tr>
<tr>
<td>--pars_tree PARS_TREE</td>
<td>--tree pars</td>
<td>Start tree option: parsimony-based randomized stepwise addition algorithm</td>
</tr>
<tr>
<td>--user_tree USER_TREE</td>
<td>--tree</td>
<td>Start tree option: upload custom tree in Newick format</td>
</tr>
<tr>
<td>--search</td>
<td>--search</td>
<td>Predefined start tree: 10 random and 10 parsimony in v0.8.0 and later</td>
</tr>
<tr>
<td>--search_1random</td>
<td>--search1</td>
<td>Predefined start tree: 1 random</td>
</tr>
<tr>
<td>--constraint_tree  CONSTRAINT_TREE</td>
<td>--tree-constraint</td>
<td>Specify topological constraint tree</td>
</tr>
<tr>
<td>--outgroup OUTGROUP</td>
<td>--outgroup</td>
<td>Outgroup to root inferred tree</td>
</tr>
<tr>
<td>--bsconverge</td>
<td>--bsconverge</td>
<td>Posteriori bootstrap convergence test</td>
</tr>
<tr>
<td>--bs_msa</td>
<td>--bsmsa</td>
<td>Bootstrap replicate alignments</td>
</tr>
<tr>
<td>--bs_trees BS_TREES</td>
<td>--bs-trees</td>
<td>Number of bootstrap trees OR autoMRE</td>
</tr>
<tr>
<td>--bs_tree_cutoff BS_TREE_CUTOFF</td>
<td>--bs-cutoff</td>
<td>Specify bootstopping cutoff value</td>
</tr>
<tr>
<td>--bs_metric BS_METRIC</td>
<td>--bs-metric</td>
<td>Compare bootstrap support values</td>
</tr>
<tr>
<td>--bootstrap</td>
<td>--bootstrap</td>
<td>Non-parametric bootstrap analysis</td>
</tr>
<tr>
<td>--check</td>
<td>--check</td>
<td>Alignment sanity check</td>
</tr>
<tr>
<td>--log LOG</td>
<td>--log</td>
<td>Options for output verbosity</td>
</tr>
<tr>
<td>--loglh</td>
<td>--loglh</td>
<td>Compute and print the likelihood of the tree(s) without optimization or generating output files</td>
</tr>
<tr>
<td>--terrace TERRACE</td>
<td>--terrace</td>
<td>Check if a tree is on a phylogenetic terrace.</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--version</td>
<td>check RAxML-NG version</td>
</tr>
<tr>
<td>--seed SEED</td>
<td>Seed for random numbers</td>
</tr>
<tr>
<td>--redo</td>
<td>Run even if there are existing files with the same name</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and  stderr) (default: False)</td>
</tr>
<tr>
<td>--logfile LOGFILE</td>
<td>Name for log file (output)</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False)</td>
</tr>
<tr>
<td>--ncpu NCPU</td>
<td>Number of CPU to use (default: 1)</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe build_tree_NG --all --seqs hp_alignments/alignment.fasta --model GTR
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../expipes/" class="btn btn-neutral float-right" title="Example Pipelines">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../hp_haplotype/" class="btn btn-neutral" title="Haplotype"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../hp_haplotype/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../expipes/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
