<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>File Input/Output - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "File Input/Output";
    var mkdocs_page_input_path = "inout.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_assemble/">Assemble</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Description</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../phylo/">Phylo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">File Input/Output</a>
    <ul class="subnav">
            
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>File Input/Output</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <table>
<thead>
<tr>
<th>Module</th>
<th>Input</th>
<th>Input Format</th>
<th>Output</th>
<th>Output File Names</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>sample_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Subampled FASTQ file(s)</td>
<td><em>sample_1.fastq</em> <br> <em>sample_2.fastq</em></td>
</tr>
<tr>
<td><strong>trim_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Trimmed FASTQ file(s) and <br> summary from Trimmomatic text file</td>
<td><em>trimmed_1.fastq</em> <br> <em>trimmed_2.fastq</em> <br> <em>trimmed_U.fastq</em> <br> <em>trimmomatic_summary.out</em></td>
</tr>
<tr>
<td><strong>join_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>FASTQ files for joined reads and unjoined reads</td>
<td><em>joined.fastq</em> <br> <em>notjoined_1.fastq</em> <br> <em>notjoined_2.fastq</em></td>
</tr>
<tr>
<td><strong>ec_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Error corrected FASTQ file(s)</td>
<td><em>corrected_1.fastq</em> <br> <em>corrected_2.fastq</em> <br> <em>corrected_U.fastq</em></td>
</tr>
<tr>
<td><strong>assemble_denovo</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>De novo assembled contigs FASTA file and <br> de novo summary text file</td>
<td><em>denovo_contigs.fna</em> <br> <em>denovo_summary.txt</em></td>
</tr>
<tr>
<td><strong>assemble_amplicons</strong></td>
<td>FASTA files and <br> GTF file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> and <br> <em>.gtf</em></td>
<td>FASTA file with assembled amplicons</td>
<td><em>amplicon_assembly.fna</em></td>
</tr>
<tr>
<td><strong>assemble_scaffold</strong></td>
<td>FASTA files</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA files with scaffolded and aligned sequences and <br> FASTA file with assembled amplicons</td>
<td><em>scaffold_aligned.fa</em> <br> <em>scaffold_assembly.fa</em> <br> <em>scaffold_imputed.fa</em> <br> <em>scaffold_padded.out</em></td>
</tr>
<tr>
<td><strong>align_reads</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>Aligned BAM file and <br> Bowtie2 alignment output summary text file</td>
<td><em>aligned.bam</em> <br> <em>aligned.bt2.out</em></td>
</tr>
<tr>
<td><strong>call_variants</strong></td>
<td>BAM and <br> FASTA file</td>
<td><em>.bam</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>VCF file with variants</td>
<td><em>variants.vcf.gz</em></td>
</tr>
<tr>
<td><strong>vcf_to_consensus</strong></td>
<td>VCF file</td>
<td><em>.vcf</em></td>
<td>FASTA file with consensus sequence</td>
<td><em>consensus.fna</em></td>
</tr>
<tr>
<td><strong>refine_assembly</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA file with refined consensus sequence</td>
<td><em>refined.fna</em></td>
</tr>
<tr>
<td><strong>finalize_assembly</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA file with final refined consensus sequence and <br> final BAM file with reads aligned to final FASTA file and <br> VCF file with variants relative to <em>final.fna</em></td>
<td><em>final.fna</em> and <br> <em>final.bam</em> and <br> <em>final.vcf.gz</em></td>
</tr>
<tr>
<td><strong>predict_haplo</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>PredictHaplo's fasta-like output file</td>
<td><em>best.fa</em></td>
</tr>
<tr>
<td><strong>ph_parser</strong></td>
<td>PredictHaplo's FAS output file</td>
<td><em>.fas</em></td>
<td>Summary text file with haplotype diversity statistics and <br> FASTA file with haplotype sequences</td>
<td><em>ph_summary.txt</em> and <br> <em>ph_haplotypes.fna</em></td>
</tr>
<tr>
<td><strong>clique_snv</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA file with haplotype sequences and TXT files with summary information</td>
<td><em>cs0X_REGION.fasta</em> <br> <em>cs0X_REGION.txt</em> <br> <em>cs0X_REGION_summary.txt</em></td>
</tr>
<tr>
<td><strong>pairwise_align</strong></td>
<td>FASTA files and <br> GTF file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> and <br> <em>.gtf</em></td>
<td>JSON file</td>
<td><em>pairwise_aligned.json</em></td>
</tr>
<tr>
<td><strong>extract_pairwise</strong></td>
<td>JSON file from pairwise_align output</td>
<td><em>.json</em></td>
<td>FASTA output with region extracted to standard out</td>
<td><em>stdout.fasta</em></td>
</tr>
<tr>
<td><strong>summary_stats</strong></td>
<td>Direcoty lists for assembly and haplotype files (TXT)</td>
<td><em>.txt</em></td>
<td>Summary statistics (TXT and TSV)</td>
<td><em>summary_stats.txt</em> <br> <em>summary_stats.tsv</em> <br> <em>PH_summary_stats.tsv</em></td>
</tr>
<tr>
<td><strong>multiple_align</strong></td>
<td>FASTA files and/or <br> directory list (TXT) <br> and GTF file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> and <em>.txt</em> <br> and <br> <em>.gtf</em></td>
<td>FASTA alignment and sequence files separated by amplicons in GTF</td>
<td><em>alignment_regionX.fasta</em>  <br> <em>all_sequences_regionX.fasta</em></td>
</tr>
<tr>
<td><strong>model_test</strong></td>
<td>FASTA or PHYLIP alignment file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> or <br> <em>.phy</em></td>
<td>Text file containing best-fit models</td>
<td><em>modeltest_results.out</em></td>
</tr>
<tr>
<td><strong>build_tree</strong></td>
<td>FASTA or PHYLIP alignment file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>RAxML output files</td>
<td><em>RaxML_info.build_tree.tre</em> and RAxML tree files</td>
</tr>
</tbody>
</table>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../demos/" class="btn btn-neutral float-right" title="Demo">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../expipes/" class="btn btn-neutral" title="Example Pipelines"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../expipes/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../demos/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
