<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Phylo - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Phylo";
    var mkdocs_page_input_path = "phylo.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_assemble/">Assemble</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Annotate</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Phylo</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#multiple_align_">multiple_align_</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../inout/">Input/Output</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demos</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Phylo</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p><strong>Phylo</strong> includes phylogenomics stages.</p>
<h3 id="multiple_align_"><em>multiple_align_</em></h3>
<p>Align consensus sequences using MAFFT (<a href="https://mafft.cbrc.jp/alignment/software/manual/manual.html">documentation</a>). Input can be a list of directories which contain <code>final.fna</code> files or a fasta file, or both (in which case the sequences in the FASTA file are combined with the <code>final.fna</code> files retreived before the alignment.
Sequences will be separated by amplicons using a supplied GTF file before alignment (unless the <code>--alignall</code> option is specified). This module may also be used to separate files by amplicons (without aligning) by specifying the <code>--fastaonly</code> option.
Alignments are by default outputted as FASTA files, although PHYLIP (<code>--phylipout</code>) or CLUSTAL (<code>--clustalout</code>) output options are also available.
Many options from MAFFT are available in this module. Please refer to the MAFFT documentation above for information about these options.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe multiple_align [MAFFT OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --dir_list &lt;TXT&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_multiple_align [MAFFT OPTIONS] [HAPHPIPE OPTIONS] --seqs &lt;FASTA&gt; --dir_list &lt;TXT&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><em>Output files:</em> alignment files in FASTA format (default), one per amplicon (or one <code>alignment.fasta</code> file if using <code>--alignall</code> option)</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqs SEQS</td>
<td>FASTA file with sequences to be aligned</td>
</tr>
<tr>
<td>--dir_list DIR_LIST</td>
<td>List of directories which include a final.fna file,one on each line</td>
</tr>
<tr>
<td>--ref_gtf REF_GTF</td>
<td>Reference GTF file</td>
</tr>
<tr>
<td>--out_align OUT_ALIGN</td>
<td>Name for alignment file</td>
</tr>
<tr>
<td>--nuc</td>
<td>Assume nucleotide (default: False)</td>
</tr>
<tr>
<td>--amino</td>
<td>Assume amino (default: False)</td>
</tr>
<tr>
<td>--clustalout</td>
<td>Clustal output format (default: False)</td>
</tr>
<tr>
<td>--phylipout</td>
<td>PHYLIP output format (default: False)</td>
</tr>
<tr>
<td>--inputorder</td>
<td>Output order same as input (default: False)</td>
</tr>
<tr>
<td>--reorder</td>
<td>Output order aligned (default: False)</td>
</tr>
<tr>
<td>--treeout</td>
<td>Guide tree is output to the input.tree file (default:False)</td>
</tr>
<tr>
<td>--quiet_mafft</td>
<td>Do not report progress (default: False)</td>
</tr>
<tr>
<td>--outdir OUTDIR</td>
<td>Output directory</td>
</tr>
</tbody>
</table>
<p><em>MAFFT Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--algo ALGO</td>
<td>Use different algorithm in command: linsi, ginsi,   einsi, fftnsi, fftns, nwns, nwnsi</td>
</tr>
<tr>
<td>--auto</td>
<td>Automatically select algorithm (default: False)</td>
</tr>
<tr>
<td>--sixmerpair</td>
<td>Calculate distance based on shared 6mers, on by default (default: False)</td>
</tr>
<tr>
<td>--globalpair</td>
<td>Use Needleman-Wunsch algorithm (default: False)</td>
</tr>
<tr>
<td>--localpair</td>
<td>Use Smith-Waterman algorithm (default: False)</td>
</tr>
<tr>
<td>--genafpair</td>
<td>Use local algorithm with generalized affine gap cost (default: False)</td>
</tr>
<tr>
<td>--fastapair</td>
<td>Use FASTA for pairwise alignment (default: False)</td>
</tr>
<tr>
<td>--weighti WEIGHTI</td>
<td>Weighting factor for consistency term</td>
</tr>
<tr>
<td>--retree RETREE</td>
<td>Number of times to build guide tree</td>
</tr>
<tr>
<td>--maxiterate MAXITERATE</td>
<td>Number of cycles for iterative refinement</td>
</tr>
<tr>
<td>--noscore</td>
<td>Do not check alignment score in iterative alignment (default: False)</td>
</tr>
<tr>
<td>--memsave</td>
<td>Use Myers-Miller algorithm (default: False)</td>
</tr>
<tr>
<td>--parttree</td>
<td>Use fast tree-building method with 6mer distance (default: False)</td>
</tr>
<tr>
<td>--dpparttree</td>
<td>Use PartTree algorithm with distances based on DP (default: False)</td>
</tr>
<tr>
<td>--fastaparttree</td>
<td>Use PartTree algorithm with distances based on FASTA (default: False)</td>
</tr>
<tr>
<td>--partsize PARTSIZE</td>
<td>Number of partitions for PartTree</td>
</tr>
<tr>
<td>--groupsize GROUPSIZE</td>
<td>Max number of sequences for PartTree</td>
</tr>
</tbody>
</table>
<p><em>MAFFT Parameters:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--lop LOP</td>
<td>Gap opening penalty</td>
</tr>
<tr>
<td>--lep LEP</td>
<td>Offset value</td>
</tr>
<tr>
<td>--lexp LEXP</td>
<td>Gap extension penalty</td>
</tr>
<tr>
<td>--LOP LOP</td>
<td>Gap opening penalty to skip alignment</td>
</tr>
<tr>
<td>--LEXP LEXP</td>
<td>Gap extension penalty to skip alignment</td>
</tr>
<tr>
<td>--bl BL</td>
<td>BLOSUM matrix: 30, 45, 62, or 80</td>
</tr>
<tr>
<td>--jtt JTT</td>
<td>JTT PAM number &gt;0</td>
</tr>
<tr>
<td>--tm TM</td>
<td>Transmembrane PAM number &gt;0</td>
</tr>
<tr>
<td>--aamatrix AAMATRIX</td>
<td>Path to user-defined AA scoring matrix</td>
</tr>
<tr>
<td>--fmodel</td>
<td>Incorporate AA/nuc composition info into scoring matrix (default: False)</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu NCPU</td>
<td>Number of CPU to use (default: 1)</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and  stderr) (default: False)</td>
</tr>
<tr>
<td>--logfile LOGFILE</td>
<td>Name for log file (output)</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False)</td>
</tr>
<tr>
<td>--fastaonly</td>
<td>Output fasta files separated by region but do not     align (default: False)</td>
</tr>
<tr>
<td>--alignall</td>
<td>Do not separate files by region, align entire file (default: False)</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe multiple_align --dir_list demo_sra_list.txt --ref_gtf HIV_B.K03455.HXB2.gtf --phylipout --logfile demo_multiple_align.log
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../expipes/" class="btn btn-neutral float-right" title="Example Pipelines">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../hp_haplotype/" class="btn btn-neutral" title="Haplotype"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../hp_haplotype/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../expipes/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
