<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Haphpipe User Guide - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Haphpipe User Guide";
    var mkdocs_page_input_path = "haphpipe_userguide.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_assemble/">Assemble</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Annotate</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../phylo/">Phylo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../inout/">File Input/Output</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demos</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Haphpipe User Guide</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="haphpipe-user-guide">Haphpipe User Guide</h1>
<p><em><strong>HA</strong>plotype and <strong>PH</strong>ylodynamics pipeline for viral assembly, population genetics, and phylodynamics.</em></p>
<p>We assume basic familiarity with conda environments, basic bash knowledge and knowledge of general next-generation sequencing (NGS) concepts. For more information regarding all of these, see helpful links and FAQ. <strong>(insert links to markdown pages once these are split up)</strong> Remember that a directory is a simply a folder nesting system, similar to what you see on your computer. Where you see "folder" below you can also say "directory."</p>
<p><img alt="directory structre" src="/Users/keyliegibson/Desktop/Working_projects/haphpipe/Things to add to github/593d447b94e14c2d4418f0f58a604169.png" /></p>
<p>HAPHPIPE is intended only for Linux and Mac OS X platforms. </p>
<p>This User Guide was developed by undergrads and graduate students to be accessible for users at all stages. Maggie Steiner, Keylie M Gibson, Matthew L Bendall, .... all contributed to developing this User Guide and testing HAPHPIPE.</p>
<p>See <strong>paper</strong> for more information and <strong>paper</strong> for validation study.
Citing HAPHPIPE:</p>
<pre><code>insert eventually
</code></pre>

<p><strong><em>I want each of these a separate document, for now I'll just make a giant one.</em></strong></p>
<hr />
<h1 id="installation">Installation</h1>
<p>For now, the installation goes like this. We are hoping to make it into a stand alone bioconda recipe soon.</p>
<p>HAPHIPE depends on more than a dozen different programs, each of which may itself depend on other programs and libraries. Installing everything separately would be a nightmare, so you should use the package manager "Bioconda" to install everything. Bioconda is a popular package manager in the bioinformatics world. See helpful links for more information and resources for Bioconda. This User Guide describes where to get Bioconda and how to install it, then how to use Bioconda to install the necessary programs for HAPHPIPE. This User Guide details the acquisition and installation of one program, GTAK, that is not handled by Bioconda, and finally the acquisition and installation of HAPHPIPE itself.</p>
<p>Here, we describe the procedure for installing HAPHPIPE using the package manager Bioconda (Gr√ºning et al. 2018) on the command line. If you are unfamiliar with Bioconda, please see https://bioconda.github.io for installation and channel setup. HAPHPIPE is available at https://github.com/gwcbi/haphpipe and is written in Python 3.7.2 coding language. An online documentation is available at https://github.com/gwcbi/haphpipe/wiki. The installation process begins with the creation of a conda environment that installs the necessary dependencies required by HAPHPIPE. Once the conda environment has been created, it can be activated for use with the command conda activate haphpipe. Due to license restrictions, Bioconda cannot distribute and install GATK (McKenna et al. 2010; Van der Auwera et al. 2013; Poplin et al. 2018) directly. To fully install GATK, you must download a licensed copy of GATK (version 3.8-0) from the Broad Institute (https://software.broadinstitute.org/gatk/download/archive). You can then install HAPHPIPE using the single command pip install git+git://github.com/gwcbi/haphpipe.git, which pulls the repository from Github.</p>
<p><need homebrew then wget></p>
<p><strong>1. Install <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda</a></strong></p>
<p>Download the conda package:</p>
<p><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
sh Miniconda3-latest-MacOSX-x86_64.sh</code></p>
<p><strong>2. Set up <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda channels</a></strong></p>
<p>These need to be put in as a command in the order they come, as this sets the priority for where packages are pulled from. Therefore, in this order, conda-forge is top priority and explored first when downloading a program.</p>
<pre><code>conda config --add channels R
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
</code></pre>

<p><strong>3. Create a conda environment with the following dependencies</strong></p>
<pre><code class="bash">conda create -n haphpipe \
    python \
    future \
    pyyaml \
    biopython \
    seqtk \
    bowtie2 \
    bwa \
    flash \
    freebayes \
    mummer \
    picard \
    trimmomatic \
    samtools=1.9 \
    gatk=3.8 \
    spades \
    blast \
    sierrapy

</code></pre>

<p><strong>4. Activate the environment</strong></p>
<pre><code>conda activate haphpipe
</code></pre>

<p><strong>5. Install GATK</strong></p>
<p>GATK can be dowloaded before or after creating the haphpipe conda environment, but it <em>must be registered after</em> creating the conda environment.</p>
<p>Due to license restrictions, bioconda cannot distribute
and install GATK directly. To fully install GATK, you must
download a licensed copy of GATK (version 3.8-0) from the Broad Institute:
<a href="https://software.broadinstitute.org/gatk/download/archive">https://software.broadinstitute.org/gatk/download/archive</a>.</p>
<p>Once you download a copy, <em>register the package using gatk3-register</em>:</p>
<pre><code>gatk3-register /path/to/GenomeAnalysisTK-3.8-0-ge9d806836.tar.bz2
</code></pre>

<p>This will copy GATK into your conda environment.</p>
<p>NOTE: HAPHPIPE was developed and tested using GATK 3.8.</p>
<p><strong>6. Install HAPHPIPE</strong></p>
<pre><code>pip install git+git://github.com/gwcbi/haphpipe.git
</code></pre>

<p>Upon completion of the installation, you can test it to ensure the repository has been installed completely and correctly by running haphpipe -h (in quick start). Once HAPHPIPE is installed and performing correctly, there is no need to install it again; simply activate the conda environment when needed by executing <code>conda activate haphpipe</code>. If a new version is released in the future, HAPHPIPE can be updated by running the command line <code>pip install --upgrade git+git://github.com/gwcbi/haphpipe.git</code>. At any point, the <code>-h</code> option that follows any HAPHPIPE stage will output a help message that provides a description of the stage and the desired input(s) and output(s).</p>
<p><strong>7. Installing PredictHaplo</strong>
The user is required to install PredictHaplo on their own, as there are system-dependent variables within the installation process. We were unable to successfully incorporate PredictHaplo into HAPHPIPE so this step was unnecessary for you to install.</p>
<p>Users are required to download PredictHaplo on their own computing system prior to running any of the haplotype stages (<em>hp_predict_haplo</em> and <em>hp_ph_parser</em>).</p>
<p>Here is how the GW CBI team installed PredictHaplo onto our HPC, which has a slurm scheduling system and uses Lua module files.</p>
<p>We cannot help with the installation of this software, but have provided the code that we used here to install PredictHaplo onto our system. Please see <a href="https://bmda.dmi.unibas.ch/software.html">their website</a> for contact information if you need help installing their program.</p>
<p>You can download the new version for paired-end reads of the PredictHaplo program from <a href="https://bmda.dmi.unibas.ch/software.html">their website</a>. The instructions are in the README in the folder downloaded. We have also provided them here for your access. <strong>provide the readme file for them</strong></p>
<p>This module loads predicthaplo onto GWU's HPC - Colonial One. See https://bmda.dmi.unibas.ch/software.html</p>
<pre><code class="bash">cd /path/to/modules/predicthaplo

# use gcc 4.9.4, add blas and lapack to library path
module load gcc/4.9.4
module load blas/gcc/64
module load lapack/gcc/64

# download source
cd archive
wget https://bmda.dmi.unibas.ch/software/PredictHaplo-Paired-0.4.tgz
cd ..

# unzip source and change directory name
tar xfvz archive/PredictHaplo-Paired-0.4.tgz
mv PredictHaplo-Paired-0.4 0.4
cd 0.4

# install scythestat
tar xfvz scythestat-1.0.3.tar.gz
cd scythestat-1.0.3
./configure --prefix=/path/to/modules/predicthaplo/0.4/NEWSCYTHE
make install
cd ..

# compile predicthaplo
make
</code></pre>

<p>If a <code>segfault</code> error occurs during the <code>hp_predict_haplo</code> stage, this is <strong>not</strong> a characteristic of HAPHPIPE but rather that of PredictHaplo. Sometimes, we have luck if we just rerun the code again or move to an interactive CPU node. We are unsure what causes this error, and we only see it between the local and global reconstruction phases in PredictHaplo.</p>
<p></br></p>
<h2 id="-">---</h2>
<h1 id="quick-start">Quick start</h1>
<h2 id="activate-haphpipe">Activate HAPHPIPE</h2>
<p><strong>1. Activate haphpipe</strong></p>
<p>Make sure you have conda running.
For students at GW using Colonial One, you need to load the <code>miniconda3</code> module like such prior to activating the haphpipe conda environemnt: 
<code>module load miniconda3</code>.</p>
<pre><code>conda activate haphpipe
</code></pre>

<p><strong>2. Test that it is loaded correctly</strong></p>
<pre><code>haphpipe -h
</code></pre>

<p>should produce:</p>
<pre><code>Program: haphpipe (haplotype and phylodynamics pipeline)
Version: 0.8.1

Commands:
 -- Reads
    sample_reads             subsample reads using seqtk
    trim_reads               trim reads using Trimmomatic
    join_reads               join reads using FLASh
    ec_reads                 error correct reads using SPAdes

 -- Assemble
    assemble_denovo          assemble reads denovo
    assemble_amplicons       assemble contigs to amplicon regions
    assemble_scaffold        assemble contigs to genome
    align_reads              align reads to reference
    call_variants            call variants
    vcf_to_consensus         create consensus sequence from VCF
    refine_assembly          iterative refinement: align - variants - consensus
    finalize_assembly        finalize consensus sequence

 -- Haplotype
    predict_haplo            assemble haplotypes with PredictHaplo
    ph_parser                parse output from PredictHaplo.

 -- Annotate
    pairwise_align           align consensus to an annotated reference
    extract_pairwise         extract sequence regions from pairwise alignment
    annotate_from_ref        annotate consensus from reference annotation

 -- Miscellaneous
    demo                     setup demo directory and test data
</code></pre>

<h2 id="demo">Demo</h2>
<p>Demo1, de novo assembly, tell them exactly what fastq read files they need to download, exactly where to find those files, and exactly where to download them.  Provide URLs and what to expect to see in the SRL archives.  These are beginners, they don't know how to find anything.  Then tell them that Demo2 you will lead them through a reference guided assembly and that they will need to download the following finished genomes sequences and exactly where to get them</p>
<p>We will now go over two ways to complete these demos: (i) automatically, which is to ultimately test if HAPHPIPE has been installed correctly and (ii) interactively, to get a feel and experience running HAPHPIPE yourself.</p>
<p>Data for both demos is included in the repository. Both samples should run quickly (less than five minutes using eight CPUs). Note: ensure PredictHaplo is installed prior to running the demos.</p>
<ol>
<li>Automatically to test the installation of HAPHPIPE</li>
</ol>
<p><code>insert the code to run the demo data</code></p>
<ol>
<li>Interactively to familiarize yourself with running HAPHPIPE</li>
</ol>
<h2 id="example-pipelines">Example pipelines</h2>
<p>The example pipelines are written in bash scripting language. The reference files used in both examples are included in the demo data. To run in haphpipe, execute one of the following lines:</p>
<p>Pipeline 1 implements amplicon assembly using a <em>de novo</em> approach. Reads are
error-corrected and used to refine the initial assembly, with up to 5
refinement steps.</p>
<p>Pipeline 2 implements amplicon assembly using a reference-based mapping 
approach. Reads are error-corrected and used to refine the initial assembly,
with up to 5 refinement steps.</p>
<h3 id="pipeline-1-haphpipe_assemble_01">Pipeline 1: <strong><code>haphpipe_assemble_01</code></strong></h3>
<p>This pipeline implements <em>de novo</em> assembly. Reads are first trimmed (<em>trim_reads</em>) and used as input for denovo assembly (<em>assemble_denovo</em>). The <em>de novo</em> assembly stage automatically performs error correction on the trimmed reads. The assembled contigs are used as input for amplicon assembly (<em>assemble_amplicons</em>) along with reference FASTA and GTF files. The assembly is then iteratively refined up to five times (<em>refine_assembly</em>) by mapping corrected reads to the assembled FASTA file and lastly finalized (<em>finalize_assembly</em>), resulting in a FASTA file with final consensus sequences, final VCF, and aligned BAM file.</p>
<p><img alt="haphpipe01" src="/Users/keyliegibson/Desktop/Working_projects/haphpipe/haphpipe.pdf" /></p>
<p>To see the input information for Pipeline 1, use the <code>-h</code> option again like so:
<code>haphpipe_assemble_01 -h</code>, and it will show the output:</p>
<pre><code>USAGE:
haphpipe_assemble_01 [read1] [read2] [reference_fasta] [reference_gtf] [samp_id] &lt;outdir&gt;

----- HAPHPIPE assembly pipeline 01 -----

This pipeline implements amplicon assembly using a denovo approach. Reads are
error-corrected and used to refine the initial assembly, with up to 5
refinement steps.

Input:
read1:             Fastq file for read 1. May be compressed (.gz)
read2:             Fastq file for read 2. May be compressed (.gz)
reference_fasta:   Reference sequence (fasta)
reference_gtf:     Amplicon regions (GTF)
samp_id:           Sample ID
outdir:            Output directory (default is [sample_dir]/haphpipe_assemble_01)
</code></pre>

<p>General command to execute pipeline 1:</p>
<pre><code>haphpipe_assemble_01 samp/read1.fq.gz samp/read2.fq.gz refs/ref.fasta refs/ref.gtf samp
</code></pre>

<p>Example command to run with demo01:</p>
<pre><code>insert
</code></pre>

<h3 id="pipeline-2-haphpipe_assemble_02">Pipeline 2: <strong><code>haphpipe_assemble_02</code></strong></h3>
<p>This pipeline implements reference-based mapping assembly. Reads are first trimmed (<em>trim_reads</em>) and error-corrected (<em>ec_reads</em>). The corrected reads are used as input for reference-based mapping assembly (<em>refine_assembly</em>) for up to five iterations. Lastly, the assembly is finalized (<em>finalize_assembly</em>) by mapping reads onto the refined reference sequence. The final output is a FASTA file with final consensus sequences, final VCF, and aligned BAM file.</p>
<p><strong>(insert Figure 2 here)</strong></p>
<p>To see the input information for Pipeline 1, use the <code>-h</code> option again like so:
<code>haphpipe_assemble_02 -h</code>, and it will show the output:</p>
<pre><code>USAGE:
haphpipe_assemble_02 [read1] [read2] [amplicons_fasta] [samp_id] &lt;outdir&gt;

----- HAPHPIPE assembly pipeline 02 -----

This pipeline implements amplicon assembly using a reference-based approach.
Reads are error-corrected and aligned to provided amplicon reference with up to
five refinement steps.

Input:
read1:             Fastq file for read 1. May be compressed (.gz)
read2:             Fastq file for read 2. May be compressed (.gz)
amplicons_fasta:   Amplicon reference sequence (fasta)
samp_id:           Sample ID
outdir:            Output directory (default is sample_dir/haphpipe_assemble_02)
</code></pre>

<p>General command to execute pipeline 1:</p>
<pre><code>haphpipe_assemble_02 samp/read1.fq.gz samp/read2.fq.gz refs/ref.fasta samp
</code></pre>

<p>Example command to run with demo01:</p>
<pre><code>insert
</code></pre>

<p></br></p>
<h2 id="-_1">---</h2>
<h1 id="haphpipe-suite">HAPHPIPE suite</h1>
<p>Each stage can be run on its own. Stages are grouped into 4 categories: <code>hp_reads</code>, <code>hp_assemble</code>, <code>hp_haplotype</code>, and <code>hp_annotate</code>.
More detailed description of command line options for each stage are available in the <a href="https://github.com/gwcbi/haphpipe/wiki">wiki</a>. To view all available stages in HAPHPIPE, run: </p>
<pre><code>haphpipe -h
</code></pre>

<p>Output will look like:</p>
<pre><code>Program: haphpipe (haplotype and phylodynamics pipeline)
Version: 0.8.1

Commands:
 -- Reads
    sample_reads             subsample reads using seqtk
    trim_reads               trim reads using Trimmomatic
    join_reads               join reads using FLASh
    ec_reads                 error correct reads using SPAdes

 -- Assemble
    assemble_denovo          assemble reads denovo
    assemble_amplicons       assemble contigs to amplicon regions
    assemble_scaffold        assemble contigs to genome
    align_reads              align reads to reference
    call_variants            call variants
    vcf_to_consensus         create consensus sequence from VCF
    refine_assembly          iterative refinement: align - variants - consensus
    finalize_assembly        finalize consensus sequence

 -- Haplotype
    predict_haplo            assemble haplotypes with PredictHaplo
    ph_parser                parse output from PredictHaplo.

 -- Annotate
    pairwise_align           align consensus to an annotated reference
    extract_pairwise         extract sequence regions from pairwise alignment
    annotate_from_ref        annotate consensus from reference annotation

 -- Miscellaneous
    demo                     setup demo directory and test data

</code></pre>

<p>HAPHPIPE consists of a suite of sub-commands under each stage that are invoked as follows:</p>
<p><code>haphpipe [stage] [sub-command] [options]</code></p>
<p>For example, to join paired end reads, one would invoke the following:</p>
<p><code>haphpipe join_reads --fq1 trimmed_1.fastq --fq2 trimmed_2.fastq</code></p>
<p>I want all stages to have a table like this detailing the options like this.
<img alt="example" src="/Users/keyliegibson/Downloads/Screen Shot 2019-09-28 at 5.51.07 PM.png" /></p>
<h2 id="hp_reads">hp_reads</h2>
<p><em>hp_reads</em> involves cleaning up the raw read sequences, as well as other processing steps. Modules to manipulate reads. </p>
<h3 id="sample_reads"><em>sample_reads</em></h3>
<p>Subsample reads using seqtk (<a href="https://github.com/lh3/seqtk">documentation</a>). Input is reads in FASTQ format. Output is sampled reads in FASTQ format. You do not have to have all read options (i.e., read1, read2 AND unpaired reads). You can have a combination of any of those.</p>
<h4 id="usage-and-option-summary">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe sample_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_sample_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>sample_1.fastq</code> <br> <code>sample_2.fastq</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--nreads</td>
<td>Number of reads to sample. If greater than the number of reads in file, all reads will be sampled.</td>
</tr>
<tr>
<td>--frac</td>
<td>Fraction of reads to sample, between 0 and 1. Each read has [frac]</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr), default is False.</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run, default is False.</td>
</tr>
</tbody>
</table>
<h4 id="example-usage">Example usage</h4>
<p>This pulls 1000 reads from these paired end files with a starting seed of 1234.</p>
<pre><code>haphpipe sample_reads --fq1 read_1.fastq --fq2 read_2.fastq --nreads 1000 --seed 1234
</code></pre>

<p>--</p>
<h3 id="trim_reads"><em>trim_reads</em></h3>
<p>Trim reads using Trimmomatic (<a href="http://www.usadellab.org/cms/?page=trimmomatic">documentation</a>). Input is reads in FASTQ format. Output is trimmed reads in FASTQ format.</p>
<h4 id="usage-and-option-summary_1">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe trim_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_trim_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> 
<code>trimmed_1.fastq</code> <br> 
<code>trimmed_2.fastq</code> <br> 
<code>trimmed_U.fastq</code> <br> 
<code>trimmomatic_summary.out</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--adapter_file</td>
<td>Adapter file.</td>
</tr>
<tr>
<td>--trimmers</td>
<td>Trim commands for trimmomatic (default: ['LEADING:3', 'TRAILING:3', 'SLIDINGWINDOW:4:15', 'MINLEN:36']).</td>
</tr>
<tr>
<td>--encoding</td>
<td>Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_1">Example usage</h4>
<p>This trims paired end read files 1 and 2.</p>
<pre><code>haphpipe trim_reads --fq1 read_1.fastq --fq2 read_2.fastq 
</code></pre>

<p>--</p>
<h3 id="join_reads"><em>join_reads</em></h3>
<p>Join reads using FLASH (<a href="https://www.ncbi.nlm.nih.gov/pubmed/21903629">paper</a>). Input is reads in FASTQ format. Output is joined reads in FASTQ format.</p>
<h4 id="usage-and-option-summary_2">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe join_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_join_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>joined.fastq</code> <br> <code>notjoined_1.fastq</code> <br> <code>notjoined_2.fastq</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_overlap</td>
<td>The minimum required overlap length between two reads to provide a confident overlap (default: 10).</td>
</tr>
<tr>
<td>--max_overlap</td>
<td>Maximum overlap length expected in approximately 90% of read pairs, longer overlaps are penalized.</td>
</tr>
<tr>
<td>--allow_outies</td>
<td>Also try combining read pairs in the "outie" orientation (default: False).</td>
</tr>
<tr>
<td>--encoding</td>
<td>Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_2">Example usage</h4>
<p>--</p>
<h3 id="ec_reads"><em>ec_reads</em></h3>
<p>Error correction using SPAdes (<a href="http://cab.spbu.ru/software/spades/">documentation</a>). Input is reads in FASTQ format. Output is error-corrected reads in FASTQ format. Remember that HAPHPIPE is intended for Illumina reads, therefore the error correction is based on Illumina sequencing errors.</p>
<h4 id="usage-and-option-summary_3">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe ec_reads [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_ec_reads [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>corrected_1.fastq</code> <br> <code>corrected_2.fastq</code> <br> <code>corrected_U.fastq</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run, default is False.</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_3">Example usage</h4>
<p>--</p>
<h2 id="hp_assemble">hp_assemble</h2>
<p>Assemble consensus sequence(s). Input reads (in FASTQ format) are assembled 
using either denovo assembly or reference-based alignment. 
Resulting consensus can be further refined.</p>
<h3 id="assemble_denovo"><em>assemble_denovo</em></h3>
<p>Assemble reads via de novo assembly using SPAdes (<a href="http://cab.spbu.ru/software/spades/">documentation</a>). Input is reads in FASTQ format. Output is contigs in FNA format.</p>
<h4 id="usage-and-option-summary_4">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_denovo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_denovo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>denovo_contigs.fna</code> <br> <code>denovo_summary.txt</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--no_error_correction</td>
<td>Do not perform error correction (default: False)</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for assembly</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling)</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_4">Example usage</h4>
<h3 id="assemble_amplicons"><em>assemble_amplicons</em></h3>
<p>Assemble contigs from de novo assembly using both a reference sequence and amplicon regions with MUMMER 3+ (<a href="http://mummer.sourceforge.net/manual/">documentation</a>). Input is contigs and reference sequence in FASTA format and amplicon regions in GTF format.</p>
<h4 id="usage-and-option-summary_5">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_amplicons [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_amplicons [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>amplicon_assembly.fna</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Scaffold Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
<tr>
<td>--padding</td>
<td>Bases to include outside reference annotation (default: 50).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_5">Example usage</h4>
<h3 id="assemble_scaffold"><em>assemble_scaffold</em></h3>
<p>Scaffold contigs against a reference sequence with MUMMER 3+ (<a href="http://mummer.sourceforge.net/manual/">documentation</a>). Input is contigs in FASTA format and reference sequence in FASTA format. Output is scaffold assembly, alligned scaffold, imputed scaffold, and padded scaffold in FASTA format.</p>
<h4 id="usage-and-option-summary_6">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_scaffold [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_scaffold [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>scaffold_aligned.fa</code> <br> <code>scaffold_assembly.fa</code> <br> <code>scaffold_imputed.fa</code> <br> <code>scaffold_padded.out</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqname</td>
<td>Name to append to scaffold sequence (default: sample01).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_6">Example usage</h4>
<h3 id="align_reads"><em>align_reads</em></h3>
<p>Map reads to reference sequence (instead of running de novo assembly) using Bowtie2 (<a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">documentation</a>) and Picard (<a href="https://broadinstitute.github.io/picard/">documentation</a>). Input is reads in FASTQ format and reference sequence in FASTA format. </p>
<h4 id="usage-and-option-summary_7">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe align_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_align_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>aligned.bam</em> <br> <em>aligned.bt2.out</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast, fast, sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local}</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
<tr>
<td>--no_realign</td>
<td>Do not realign indels (default: False).</td>
</tr>
<tr>
<td>--remove_duplicates</td>
<td>Remove duplicates from final alignment. Otherwise duplicates are marked but not removed (default: False).</td>
</tr>
<tr>
<td>--encoding</td>
<td>{Phred+33,Phred+64} Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_7">Example usage</h4>
<h3 id="call_variants"><em>call_variants</em></h3>
<p>Variant calling from alignment using GATK (<a href="https://software.broadinstitute.org/gatk/download/archive">documentation</a>). Input is alignment file in BAM format and reference sequence in FASTA format (either reference from reference-based assembly or consensus final sequence from de novo assembly). Output is a Variant Call File (VCF) format file. </p>
<h4 id="usage-and-option-summary_8">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe call_variants [OPTIONS] [SETTINGS] --aln_bam &lt;BAM&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_call_variants [OPTIONS] [SETTINGS] --aln_bam &lt;BAM&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>variants.vcf.gz</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--aln_bam</td>
<td>Alignment file.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--emit_all</td>
<td>Output calls for all site (default: False).</td>
</tr>
<tr>
<td>--min_base_qual</td>
<td>Minimum base quality required to consider a base for calling (default: 15).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_8">Example usage</h4>
<h3 id="vcf_to_consensus"><em>vcf_to_consensus</em></h3>
<p>Generate a consensus sequence from a VCF file. Input is a VCF file. Output is the consensus sequence in FASTA format. </p>
<h4 id="usage-and-option-summary_9">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe vcf_to_consensus [OPTIONS] [SETTINGS] --vcf &lt;FASTQ&gt; [--outdir] [--sampidx]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_vcf_to_consensus [OPTIONS] [SETTINGS] --vcf &lt;FASTQ&gt; [--outdir] [--sampidx]</code></p>
<p><em>Output files:</em> <br> <em>consensus.fna</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--vcf</td>
<td>VCF file (created with all sites).</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
<tr>
<td>--sampidx</td>
<td>Index for sample if multi-sample VCF (default: 0).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_DP</td>
<td>Minimum depth to call site (default: 1).</td>
</tr>
<tr>
<td>--major</td>
<td>Allele fraction to make unambiguous call (default: 0.5).</td>
</tr>
<tr>
<td>--minor</td>
<td>Allele fraction to make ambiguous call (default: 0.2).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_9">Example usage</h4>
<h3 id="refine_assembly"><em>refine_assembly</em></h3>
<p>Map reads to a denovo assembly or reference alignment. Assembly or alignment is iteratively updated. Input is reads in FASTQ format and reference sequence (assembly or reference alignment) in FASTA format. Output is refined assembly in FASTA format.</p>
<h4 id="usage-and-option-summary_10">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe refine_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_refine_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>refined.fna</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--max_step</td>
<td>Maximum number of refinement steps (default: 1).</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for refinement.</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_10">Example usage</h4>
<h3 id="finalize_assembly"><em>finalize_assembly</em></h3>
<p>Finalize consensus, map reads to consensus, and call variants. Input is reads in FASTQ format and reference sequence in FASTA format. Output is finalized reference sequence, alignment, and variants (in FASTA, BAM, and VCF formats, respectively).</p>
<h4 id="usage-and-option-summary_11">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe finalize_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_finalize_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> final.fna <br> final.ban <br> final.vcf.gz</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Consensus fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast,fast,sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local} Bowtie2 preset to use (default: very-sensitive).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_11">Example usage</h4>
<h2 id="hp_haplotype">hp_haplotype</h2>
<p>Haplotype assembly stages. HAPHPIPE implements PredictHaplo (<a href="https://www.ncbi.nlm.nih.gov/pubmed/26355517">paper</a>), although other haplotype reconstruction programs can be utilized outside of HAPHPIPE using the final output of HAPHPIPE, typically with the final consensus sequence (FASTA) file, reads (raw, trimmed, and/or corrected), and/or final alignment (BAM) file as input.</p>
<h3 id="predict_haplo"><em>predict_haplo</em></h3>
<p>Haplotype identification with PredictHaplo. Input is reads in FASTQ format and and reference sequence in FASTA format. Output is the longest global haplotype file and corresponding HTML file. <em>Note: PredictHaplo must be installed separately before running this stage.</em> </p>
<h4 id="usage-and-option-summary_12">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe predict_haplo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; --interval_txt [TXT] [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_ predict_haplo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; --interval_txt [TXT] [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>best.fa</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference sequence used to align reads (Fasta).</td>
</tr>
<tr>
<td>--interval_txt</td>
<td>File with intervals to perform haplotype reconstruction.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_readlength</td>
<td>Minimum read length passed to PredictHaplo (default: 36).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_12">Example usage</h4>
<h3 id="ph_parser"><em>ph_parser</em></h3>
<p>Returns PredictHaplo output as a correctly formatted FASTA file. Input is the output file from predict_haplo (longest global .fas file). Output is a correctly formatted FASTA file.</p>
<h4 id="usage-and-option-summary_13">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe ph_parser [OPTIONS] [SETTINGS] --haplotypes_fa &lt;best.fas&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_ph_parser [OPTIONS] [SETTINGS] --haplotypes_fa &lt;best.fas&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>ph_summary.txt</em> <br> <em>ph_haplotypes.fna</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--haplotypes_fa</td>
<td>Haplotype file created by PredictHaplo.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--prefix</td>
<td>Prefix to add to sequence names.</td>
</tr>
<tr>
<td>--keep_gaps</td>
<td>Do not remove gaps from alignment (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_13">Example usage</h4>
<p>By default, PredictHaplo outputs their own unique version of a fasta file. It includes the frequency, some information regarding their unqiue overlapping scores, and their unique confidence scores. This file is always named <code>PH#.best_#_#.fas</code>, where the first number is the reconstructed haplotype number and the next numbers are the start and end of the longest haplotype reconstructed by PredictHaplo.</p>
<pre><code>$ cat PH01.best_1_1208.fas
&gt;reconstructed_0
;Freq:0.190638
;Overlap quality scores (5,10):1,1
;Confidence scores
;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~n7~~~y~~~~~t~~~i~jjk~~zz
;~~~~{~~~~~N{~{~Sx~~~~~~z~K~F~~~~~~~y~~~~~~~|~~F~wx|~~{~~|~~s&lt;|]~~~~~~
;~m~kj|{|v~{|_`~~~z~~~~~~~jy{y~~~~~a~__~~|~~~~~{wXZ~}~~~qm~xV~~~~~~~}~
;Q~}~y||~~~}}~~~z~~~~~~{}A~}b|~~u~~|}}|~}}z~}bx~~n||~~||{~}~~d}~bz~~~}
;|~}}~~~}~~~{|}}g{~~~}~r~}}~~~~u~~{kx{~}~}~|~}~}{}~}~}||~~~~[~}~}}~~~~
;~~}~~~U||U}~|}}~~}}~~~~}u~b|}~w~~~~~{}|wv~}Dxzp{|}~@~~P~}~}~V~~z}~|ry
;q~}|}~}~~}t~o~}~f}~~}{~~~}~{~~~~~~~}~~}~~|~~~M~~}~}x~}c~}^v~~~yzA~}~}
;y}~z}~~~~~~~~~{z~~}~~~}~{}~~~~~}~~~~~~~|~~v~~}~~|y~]|{~||~~~~~~~~||~~
;Y||~~|Q~|~~~|~~~~~|~~~z~~z~{{{y~~~~~~~~~~w{{~wz|~Z|~z|~~}p|~~|}}~~x}}
;z}~}}|a|}}}}{}|~~}}~}}~}~{~|~}}~}{}|}|}}}~|}}}}{}}}|}}|}|}}}|}}}}~}}}
;||}}}}{}}~}~}}}}}}}}}}}~}}}}}}}}}}}}}}}}}y}~}}}}}}}}}}~|}}}}|}}}}|}}}
;}}}}}}}}|}}}}}}}}~}}}}}}|}}}}}}}}|}}}}}|}}}}|~|}}}}{}}}}}}}}}}}}}}}}}
;}}=}}}{}}}}}}}}}}}}}}}}}}}}}}}}~||}}|~}}}}}{}}}}}|}}|}}}}}}}}}}}}}|}}
;|}~}}}}|}}}}}}}}|}|~J}}}}}}~}}}}}}}}}}}}}}}}}}}}|}}`~}}}}|}}}}}}}}}}~
;|}}}}}}}}}}}~}}}|}}}}}}}}}}|}}}}}}}}}}}t}}{}`}}}}}}|}~}~}|~}}}|k}}}}}
;|}|~}}|}}~}}~}|}}z}}}}}}}}}}}}}}~}}~}}}~~}|}}}~}}}}}}}}~}~}}||~~}~z}}
;~~}~}||~|~}|{}||~|z~~||}~|~}}|}~~}|}}~}}|}z~~~~{~}{}~y~~~~~{|}}~~|y~~
;~|~~||~~~~~~~|n~~{~~~~~~~~~~~~~~~
;EndOfComments
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAATATCTGGCCTTCCCGCAAGGGAAG
GCCAGGGAACTTTCCTCAGAGCAGGCTAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGCTTGGGGAA
GCAACAACAAAGCCCTCTCAGAAACAGGAGCCAATAGACAAGGAAATGTATCCTTTAACTTCCCTCAGAT
CACTCTTTGGCAACGACCCCTTGTCACAATAAAGATAGGGGAGCAACTAAAGGAAGCCCTGTTAGATACA
GGAGCAGATGATACAGTATTAGAAGAAATGAATTTGCCAGGAAGATGGAAACCAAGAATGATAGGGGGAA
TCGGGGGTTTTATCAAAGTAAGACAGTATGATCAGATAGCCATAGACATCTGTGGCCATAAAGCTATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGCTTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTGAAGCCAGGAATGGATGGCCCAAAGG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAGTAGAAATTTGTACAGAAATGGAAAAAGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGGAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAGAGAACTTAATAAGAGAACTCAGGACTTCTGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAATCAGTAACAGTACTGGATGCGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTATACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACTAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AAGCTAGCATGACAAAAATCTTAGAGCCTTTTAGAAAACAAAATCCAGACATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCTT
&gt;reconstructed_1
;Freq:0.294104
;Overlap quality scores (5,10):1,1
;Confidence scores
;~~~~~~~~z~~~~~~~~~|~y~~~~u}|~~~~~}}~~~~~~~ya~|~~}}~~~y~~~j~YXH~~s~
;~~~~z~~|~~b|}^}xZ}}~}z~t~r~{}~}x}}yb}~~~}~}u~~}~gu|}}{~|}}~ozsw}|}}h~
;|l}v|^][t}zz}vw}}s}}}}~v}|~|~~}}}~~}}}}}~}}}}}}zo\}}}~}vn}iv}}}}}}}}}
;v}}~|u}}}}}}}}}|}}}}}}|}^}}[}~}r}}}{|}}|t|}}{z}}pz}}{}}}}}}}v~}y|~}}}
;}}}}~}}~}}}tz}}`}}}~}}q}}}}}~}y}}|r||}}}}~}}~}}{}}}}}}}}}}}y~}}}}t}}}
;}}}}}}{|}|}}{r}}}}tv~}}}t}m|~}h}}}}}v}}qt}|y|{|u~}}|}~_}~|~~|}~|}~}|e
;Y}~}}~}}~}v}t}}}j}}~}}}~}}~|}}}}|}}|}}}~}}}}};}}~}}x}}|}}yu}}}wy`}|~~
;{}}x}}}}}~}}}~{v}}}}}{}~}}~}~}}|}}}}}}}}}}u}}y~~{g}L}}}}}|~}{~|}|}|}}
;l}}}}{_}}}}~z~}|}~~|~~x|~}~z~~|~}~~~}~~~}y}~~|k}}Pf}w~~~~T}}}F~~}~z}{
;{~|}|zZ~s~|}r~}}~~|}}}}}~|}}}}n}}}~}}}}~}}}}}}}}}}}|~}}}}}}~~}}~~~~~}
;N}}}}}|~}}~}}}}{}}}~}}|~~|}|}}}~|}~}}}}}}D}~}}}}}}}}}}}}}}}}|}}z}|}}}
;}k}}}}}}}}}}}}}}}}}}}|}}}}}|}}}}}|}}}}}|}}}}|~{~}}}G}~}}}}}~}}}}}}}}}
;}}h}}~|}|}}}}}}}}}}}}}}}}}}}}}}}}{}}|}}}}|~|}}}}}}}M}}}}}}}~}}}}}}}}}
;}}~|}}}}}~~~}}}~}}|}g}}}}}}}}~|}}}}}}}}}}}}}}}}}|}~t~}|}}}}}}}||~}}}}
;|}}}}~}}}}}}~}}}gi}}~}}}}}}}}}}}}|}}~}}g}~}}v}}}}}|{}}}}}}~}}~|e}}}}}
;}}|}}~}~}~}}~}|~}|}~|||}}}}|~}}~}}}}}}}}|}}~}}~}|~}|y~}~}}}}}|}}~~}~}
;}}~}~}|~}}}~}}}}~}{}}~}}~}}}||~~}||~}x|}{~|}|~~|~}}}|{~}}~}{}}||~}}}~
;|~}~}}}~|}}~~|,~~~~~~~~~~~~~~m~~~
;EndOfComments
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAAGATCTGGCCTTCCCGCGGCGGAAG
GCCAGGGAATTTTCTTCAGAGCAGACCAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGTTTGGGGAG
GAGACAACAACTCCCTCTCAGAAGCAGGAGCCGAAGGACAAGGAAACATATCCTTTAGCTTCCCTCAAAT
CACTCTTTGGCAACGACCCCTTGTTACAATAAAGATAGGGGGGCAGCTAAAGGAAGCTCTATTAGATACA
GGAGCAGATGACACAGTATTAGAAGAAATGGATTTGCCAGGAAGATGGAAACCAAAAATGATAGGGGGAA
TTGGAGGTTTTATCAAAGTAAAACAGTATGATCAGATACCCATAGAAATTTGTGGACATAAAGTGATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGCTTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTGAAGCCAGGAATGGATGGCCCAAAGG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAATAGAAATCTGTGCAGAAATGGAAAAGGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGAAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAAAGAACTTAATAAGAGAACTCAGGACTTCTGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAGTCAGTAACAGTACTGGATGTGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTACACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACCAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AATGTAGCATGACAAAAATCTTAGAACCTTTTAGAAAACAAAATCCAGATATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCTT
&gt;reconstructed_2
;Freq:0.515259
;Overlap quality scores (5,10):4,4
;Confidence scores
;~~~~~~~w~~zz~~~~z~~w~|wy~|~~{|~~|~~~~~~~~}~{~hD~~~}|~|}}t~~_;~tue}}S}
;~}~~p}}l}~P}}ZlvL~z}}h~g~A~x}}}s}}{S~}~}}}}l}~|}SOq}}q}w}}~i\pz~w~~z~
;}7~|vzUu4|smtgm}~W}}~~}r}n`f`}{PZLSI:Vs_V_}}}}_&lt;db}}}}}CQ}YJ}t}}}~}}}
;T}}~|}}}}}}}~}}k}}}}}}q}L}}]}}}L}}}tt|~}{q}}{M}lAIu}}}s{}}}~y~}st}|}}
;z}}}}|}}}}}{q|}Z}}}}}}P~}|}|}}E}}cLbi}vu}}|}}|}g}}}}}}|}~}~{}~}|}{}}}
;}~}|}}}{{|~}pp}~~~{|}}~}[yEx}}\}}}{}N}uoK}}pauSy}}}v~~J}}q}}v~}c}~}\i
;]|}}~}|}~}r}Y}}}B}}~||~}}}}{}}}|}|~|}~~~~}}~{w}{}~|R|}s|}k{~}}}ra}q}}
;o}}d}~}}}}}~}}|X~}}~}e}}}}}}}|}y|}}}}~}|~}c~~b~~nL~a}}}}}~}}m}}~{}}}}
;g}}}}uLm~l}}`~|n}}~}}|d~~y~u}|tz~~~~|}}~~t}}}tx}|Tu~v}}~}^|~~w~|~}m{}
;u}}}|w;}{}{}r}||}}z|}|~|}z}||}y}}}~}}}|}}~}}}|}y}}}}}}}~}||}}}}}}~}}|
;v{}|}}{}}}|~|}}|}}}}}}}}}}|}}}}}|}~}}}}}}p}}||}}}}}~}}}{}}}}}}}z}}}}}
;}|}}~}}}|}}}}}}}}}}}}}}}}}}}}}}|}||}}|}|}}}}|}|}}}}u}}}}}}}}}}}}}}}}}
;}}g}}}{}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}y}}}}}|}m}}}}}}}}}}}}}}}}}
;}}}}}}}}}}}}}}}}}}|}i}}}}}}}}}}|}}}}}}}}}}}}}}}}}}}H}}}}}}}}}|}{}}}}}
;|}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}y}}|}H}}}}}}}}}}}}}}}}}}u}}}}}
;}}}}}}}}}}}}}}}}}z}}}}}}}}}}}}}}}}|}}}|}}}}}}}}}|}}}}}}}}}}}}|}}}}|}}
;}}}}}}}}}}}}||}|}}|}}}}}}|}}|}}}~}|}|}}}|}{}}}}|}}}}|{|}}}}x||}}}|{}}
;}}}}}}}~}}~}~|n}~}}}}}~}}~~|~~~y~,
;EndOfComments
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAAAATCTGGCCTTCCCACAAGGGAAG
GCCAGGGAATTTTCTTCAGAGCAGACCAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGTTTGGGGAA
GAGACAACAACTCCCTCTCAGAAGCAGGAGCCGATAGACAAGGAAATGTATCCTTTAGCTTCCCTCAAAT
CACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAGGGGGGCAACTAAAGGAAGCTCTATTAGATACA
GGAGCAGATGATACAGTATTAGAAGAAATGAATTTGCCAGGAAGATGGAAACCAAAAATGATAGGGGGAA
TTGGAGGTTTTATCAAAGTAAAACAGTATGATCAGATACCCATAGAAATCTGTGGACATAAAGCTATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGATTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTAAAGCCAGGAATGGATGGCCCAAAAG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAGTAGAAATTTGTACAGAAATGGAAAAGGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGGAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAGAGAACTTAATAAGAGAACTCAAGACTTCCGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAATCAGTAACAGTACTGGATGTGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTATACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACCAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AAGCTAGCATGACAAAAATCTTAGAGCCTTTTAGAAAACAAAATCCAGACATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCCC
</code></pre>

<p><em>ph_parser</em> takes this output and creates a proper fasta file with each resontructed haplotype and a text file that has the hpalotype diversity estimate.</p>
<pre><code>$ cat ph_summary.txt
PH_num_hap 3
PH_hap_diversity 0.611668153059
PH_seq_len 1208


$ cat ph_haplotypes.fna
&gt;reconstructed_0 Freq=0.190638
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAATATCTGGCCTTCCCGCAAGGGAAG
GCCAGGGAACTTTCCTCAGAGCAGGCTAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGCTTGGGGAA
GCAACAACAAAGCCCTCTCAGAAACAGGAGCCAATAGACAAGGAAATGTATCCTTTAACTTCCCTCAGAT
CACTCTTTGGCAACGACCCCTTGTCACAATAAAGATAGGGGAGCAACTAAAGGAAGCCCTGTTAGATACA
GGAGCAGATGATACAGTATTAGAAGAAATGAATTTGCCAGGAAGATGGAAACCAAGAATGATAGGGGGAA
TCGGGGGTTTTATCAAAGTAAGACAGTATGATCAGATAGCCATAGACATCTGTGGCCATAAAGCTATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGCTTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTGAAGCCAGGAATGGATGGCCCAAAGG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAGTAGAAATTTGTACAGAAATGGAAAAAGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGGAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAGAGAACTTAATAAGAGAACTCAGGACTTCTGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAATCAGTAACAGTACTGGATGCGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTATACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACTAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AAGCTAGCATGACAAAAATCTTAGAGCCTTTTAGAAAACAAAATCCAGACATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCTT
&gt;reconstructed_1 Freq=0.294104
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAAGATCTGGCCTTCCCGCGGCGGAAG
GCCAGGGAATTTTCTTCAGAGCAGACCAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGTTTGGGGAG
GAGACAACAACTCCCTCTCAGAAGCAGGAGCCGAAGGACAAGGAAACATATCCTTTAGCTTCCCTCAAAT
CACTCTTTGGCAACGACCCCTTGTTACAATAAAGATAGGGGGGCAGCTAAAGGAAGCTCTATTAGATACA
GGAGCAGATGACACAGTATTAGAAGAAATGGATTTGCCAGGAAGATGGAAACCAAAAATGATAGGGGGAA
TTGGAGGTTTTATCAAAGTAAAACAGTATGATCAGATACCCATAGAAATTTGTGGACATAAAGTGATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGCTTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTGAAGCCAGGAATGGATGGCCCAAAGG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAATAGAAATCTGTGCAGAAATGGAAAAGGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGAAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAAAGAACTTAATAAGAGAACTCAGGACTTCTGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAGTCAGTAACAGTACTGGATGTGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTACACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACCAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AATGTAGCATGACAAAAATCTTAGAACCTTTTAGAAAACAAAATCCAGATATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCTT
&gt;reconstructed_2 Freq=0.515259
ACCAAATGAAAGATTGTACTGAGAGACAGGCTAATTTTTTAGGGAAAATCTGGCCTTCCCACAAGGGAAG
GCCAGGGAATTTTCTTCAGAGCAGACCAGAGCCAACAGCCCCACCAGAAGAGAGCTTCAGGTTTGGGGAA
GAGACAACAACTCCCTCTCAGAAGCAGGAGCCGATAGACAAGGAAATGTATCCTTTAGCTTCCCTCAAAT
CACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAGGGGGGCAACTAAAGGAAGCTCTATTAGATACA
GGAGCAGATGATACAGTATTAGAAGAAATGAATTTGCCAGGAAGATGGAAACCAAAAATGATAGGGGGAA
TTGGAGGTTTTATCAAAGTAAAACAGTATGATCAGATACCCATAGAAATCTGTGGACATAAAGCTATAGG
TACAGTATTAGTAGGACCTACACCTGTCAACATAATTGGAAGAAATCTGTTGACTCAGATTGGTTGCACT
TTAAATTTTCCCATTAGTCCTATTGAAACTGTACCAGTAAAATTAAAGCCAGGAATGGATGGCCCAAAAG
TTAAACAATGGCCATTGACAGAAGAAAAAATAAAAGCATTAGTAGAAATTTGTACAGAAATGGAAAAGGA
AGGGAAAATTTCAAAAATTGGGCCTGAAAATCCATACAATACTCCAGTATTTGCCATAAGGAAAAAAGAC
AGTACTAAATGGAGAAAATTAGTAGATTTCAGAGAACTTAATAAGAGAACTCAAGACTTCCGGGAAGTCC
AATTAGGAATACCACATCCCTCAGGGTTAAAAAAGAAAAAATCAGTAACAGTACTGGATGTGGGTGATGC
ATATTTTTCAGTTCCCTTAGATGAAGACTTCAGAAAGTATACTGCATTTACCATACCTAGTGTAAACAAT
GAGACACCAGGGATTAGGTATCAGTACAATGTGCTTCCACAAGGATGGAAAGGATCACCAGCAATATTCC
AAGCTAGCATGACAAAAATCTTAGAGCCTTTTAGAAAACAAAATCCAGACATAGTTATCTATCAATACAT
GGATGATCTGTATGTAGGATCTGACCTAGAAATAGGGCAGCATAGAACAAAAATAGAGGAACTGAGAGAA
CATCTGTTGAGGTGGGGGTTTTGCACACCAGACAAGAAACATCAGAAGGAACCTCCATTCCTTTGGATGG
GTTATGAACTCCATCCCC
</code></pre>

<h2 id="hp_annotate">hp_annotate</h2>
<h3 id="pairwise_align"><em>pairwise_align</em></h3>
<p>Apply correct coordinate system to final sequence(s) to facilitate downstream analyses. Input is the final sequence file in FASTA format, a reference sequence in FASTA format, and a reference GFT file. Output is a JSON file to be used in <em>extract_pairwise</em>.</p>
<h4 id="usage-and-option-summary_14">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe pairwise_align [SETTINGS] --amplicons_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_pairwise_align [SETTINGS] --amplicons_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> 
pairwise_aligned.json</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--amplicons_fa</td>
<td>Fasta file with assembled amplicons.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions. Primary and alternate coding regions should be provided in the attribute field (for amino acid alignment).</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_14">Example usage</h4>
<h3 id="extract_pairwise"><em>extract_pairwise</em></h3>
<p>Extract sequence regions from the pairwise alignment produced in <em>pairwise_align</em>. Input is the JSON file from <em>pairwise_align</em>. Output is either an unaligned nucleotide FASTA file, an aligned nucleotide FASTA file, an amino acid FASTA file, an amplicon GTF file, or a tab-separated values (TSV) file (default: nucleotide FASTA with regions of interest from GTF file used in <em>pairwise_align</em>). </p>
<h4 id="usage-and-option-summary_15">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe extract_pairwise [OPTIONS] [SETTINGS] --align_json &lt;JSON&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_extract_pairwise [OPTIONS] [SETTINGS] --align_json &lt;JSON&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> 
stdout.fasta</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--align_json</td>
<td>JSON file describing alignment (output of <em>pairwise_align</em> stage).</td>
</tr>
<tr>
<td>--outfile</td>
<td>Output file (default: stdout).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--outfmt</td>
<td>Format for output: nuc_fa, aln_fa, amp_gtf, ost, or prot_fa (default: nuc_fa).</td>
</tr>
<tr>
<td>--refreg</td>
<td>Reference region. String format is ref:start-stop. For example, the region string to extract pol when aligned to HXB2 is HIV_B.K03455.HXB2:2085-5096.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h4 id="example-usage_15">Example usage</h4>
<h3 id="annotate_-from_ref">* annotate_ from_ref*</h3>
<p>Annotate consensus sequence from reference annotation. Input is JSON file from <em>pairwise_align</em> and reference GTF file. </p>
<h4 id="usage-and-option-summary_16">Usage and option summary</h4>
<p><strong>Usage:</strong></p>
<p><code>haphpipe annotate_from_ref [OPTIONS] [SETTINGS] --haplotypes_fa &lt;best.fas&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_annotate_from_ref [OPTIONS] [SETTINGS] --haplotypes_fa &lt;best.fas&gt; [--outdir]</code></p>
<h4 id="example-usage_16">Example usage</h4>
<p></br></p>
<h2 id="-_2">---</h2>
<h1 id="example-usage_17">Example usage</h1>
<h2 id="hp_assemble_01"><code>hp_assemble_01</code></h2>
<h2 id="hp_assemble_02"><code>hp_assemble_02</code></h2>
<h2 id="hiv">HIV</h2>
<p>Data from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA486832">NCBI bioproject: PRJNA486832</a> with <a href="https://www.biorxiv.org/content/biorxiv/early/2018/09/12/414995.full.pdf">associated article</a>. Brief description of the paper and data: <strong>insert from students</strong></p>
<h3 id="quick-start_1">Quick start</h3>
<h4 id="miniconda-and-bioconda">Miniconda and Bioconda</h4>
<p>Step 1. Installing <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda</a>.</p>
<p><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
sh Miniconda3-latest-MacOSX-x86_64.sh</code></p>
<p>Step 2. Setting up <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda channels</a>.</p>
<pre><code>conda config --add channels R
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
</code></pre>

<h4 id="downloading-data">Downloading data</h4>
<p>This is showing how to load miniconda on GW's HPC, Colonial One, which is how undergrads in CBI work on data. </p>
<pre><code>module use /groups/cbi/shared/modulefiles
module load miniconda3/4.3.27.1
</code></pre>

<p>Creating a new conda environment to pull data from SRA.</p>
<pre><code>conda create -n fastqdump sra-tools
conda activate fastqdump
</code></pre>

<p>Downloading SRA data.
This pulls the first 20 accessions from the SRA list, and only pulls 20,000 reads from SRA data. We pulled the first 20 in order, because thier paper says that between 22-24 samples were put on a single MiSeq run.</p>
<p>This code says for each accession in the accession list, sort them in order, and then take the first 20 accession numbers. For each of these 20 accession numbers, pull 20,000 reads from the SRA data and put the output read files into the accession folder. Then deactive the conda environment used to pull these SRA data.</p>
<pre><code>for acc in $(cat SraAccList.txt | sort | head -n 20); do
    mkdir -p $acc
    fastq-dump --outdir $acc --split-files -N 10001 -X 30000 $acc
done

# deactivate the conda environment now that you are done downloading data
conda deactivate
</code></pre>

<h4 id="installing-haphpipe">Installing HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h4 id="activating-haphpipe">Activating HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h3 id="haphpipe-suite_1">HAPHPIPE suite</h3>
<p><strong>insert from students</strong></p>
<h4 id="hp_reads_1">hp_reads</h4>
<pre><code>    sample_reads
    trim_reads
    join_reads
    ec_reads
</code></pre>
<h4 id="hp_assemble_1">hp_assemble</h4>
<pre><code>    assemble_denovo
    assemble_amplicons
    assemble_scaffold
    align_reads
    call_variants
    vcf_to_consensus
    refine_assembly
    finalize_assembly
</code></pre>
<h4 id="hp_haplotype_1">hp_haplotype</h4>
<pre><code>    predict_haplo
    ph_parser
</code></pre>
<h4 id="hp_annotate_1">hp_annotate</h4>
<pre><code>    pairwise_align
    extract_pairwise
    annotate_from_ref
</code></pre>
<h3 id="example-usage_18">Example usage</h3>
<h4 id="your-virus-with-both-pipelines-document-all-code-and-explanation">Your virus with both pipelines. Document all code and explanation.</h4>
<h5 id="starting-directory-structure">Starting directory structure</h5>
<p><strong>insert from students</strong></p>
<h5 id="ending-directory-structure">Ending directory structure</h5>
<p><strong>insert from students</strong></p>
<h3 id="helpful-sources">Helpful sources</h3>
<pre><code>bash scripting
conda
</code></pre>
<h3 id="faq">FAQ</h3>
<h2 id="hcv">HCV</h2>
<p>Data from <a href="https://www.ncbi.nlm.nih.gov/bioproject?LinkName=sra_bioproject&amp;from_uid=651945">NCBI bioproject: PRJNA237978</a> with <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4067308">associated article</a>. Brief description of the paper and data: <strong>insert from students</strong></p>
<h3 id="quick-start_2">Quick start</h3>
<h4 id="miniconda-and-bioconda_1">Miniconda and Bioconda</h4>
<p>Step 1. Installing <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda</a>.</p>
<p><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
sh Miniconda3-latest-MacOSX-x86_64.sh</code></p>
<p>Step 2. Setting up <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda channels</a>.</p>
<pre><code>conda config --add channels R
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
</code></pre>

<h4 id="downloading-data_1">Downloading data</h4>
<p>This is showing how to load miniconda on GW's HPC, Colonial One, which is how undergrads in CBI work on data. </p>
<pre><code>module use /groups/cbi/shared/modulefiles
module load miniconda3/4.3.27.1
</code></pre>

<p>Creating a new conda environment to pull data from SRA.</p>
<pre><code>conda create -n fastqdump sra-tools
conda activate fastqdump
</code></pre>

<p>Downloading SRA data.
This pulls the all the accessions (23) from the SRA list, and only pulls 20,000 reads from SRA data. </p>
<p>This code says for each accession in the accession list, pull 20,000 reads from the SRA data and put the output read files into the accession folder. Then deactive the conda environment used to pull these SRA data.</p>
<pre><code>for acc in $(cat SraAccList.txt); do
    mkdir -p $acc
    fastq-dump --outdir $acc --split-files -N 10001 -X 30000 $acc
done

# deactivate the conda environment now that you are done downloading data
conda deactivate
</code></pre>

<h4 id="installing-haphpipe_1">Installing HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h4 id="activating-haphpipe_1">Activating HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h3 id="haphpipe-suite_2">HAPHPIPE suite</h3>
<p><strong>insert from students</strong></p>
<h4 id="hp_reads_2">hp_reads</h4>
<pre><code>    sample_reads
    trim_reads
    join_reads
    ec_reads
</code></pre>
<h4 id="hp_assemble_2">hp_assemble</h4>
<pre><code>    assemble_denovo
    assemble_amplicons
    assemble_scaffold
    align_reads
    call_variants
    vcf_to_consensus
    refine_assembly
    finalize_assembly
</code></pre>
<h4 id="hp_haplotype_2">hp_haplotype</h4>
<pre><code>    predict_haplo
    ph_parser
</code></pre>
<h4 id="hp_annotate_2">hp_annotate</h4>
<pre><code>    pairwise_align
    extract_pairwise
    annotate_from_ref
</code></pre>
<h3 id="example-usage_19">Example usage</h3>
<h4 id="your-virus-with-both-pipelines-document-all-code-and-explanation_1">Your virus with both pipelines. Document all code and explanation.</h4>
<h5 id="starting-directory-structure_1">Starting directory structure</h5>
<p><strong>insert from students</strong></p>
<h5 id="ending-directory-structure_1">Ending directory structure</h5>
<p><strong>insert from students</strong></p>
<h3 id="helpful-sources_1">Helpful sources</h3>
<pre><code>bash scripting
conda
**insert from students**
</code></pre>
<h3 id="faq_1">FAQ</h3>
<p><strong>insert from students</strong></p>
<h2 id="influenza">Influenza</h2>
<p>Data from <a href="https://www.ncbi.nlm.nih.gov/bioproject/506454">NCBI bioproject: PRJNA506454</a> with <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4731279">associated article</a>. Brief description of the paper and data: <strong>insert from students</strong></p>
<h3 id="quick-start_3">Quick start</h3>
<h4 id="miniconda-and-bioconda_2">Miniconda and Bioconda</h4>
<p>Step 1. Installing <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda</a>.</p>
<p><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
sh Miniconda3-latest-MacOSX-x86_64.sh</code></p>
<p>Step 2. Setting up <a href="https://bioconda.github.io/user/install.html#set-up-channels">conda channels</a>.</p>
<pre><code>conda config --add channels R
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
</code></pre>

<h4 id="downloading-data_2">Downloading data</h4>
<p>This is showing how to load miniconda on GW's HPC, Colonial One, which is how undergrads in CBI work on data. </p>
<pre><code>module use /groups/cbi/shared/modulefiles
module load miniconda3/4.3.27.1
</code></pre>

<p>Creating a new conda environment to pull data from SRA.</p>
<pre><code>conda create -n fastqdump sra-tools
conda activate fastqdump
</code></pre>

<p>Downloading SRA data.
This pulls the 20 random accessions from the SRA list, and only pulls 20,000 reads from SRA data. We pulled randomly, because the study had two sets of influenza types sequenced, and we wanted to pull from both, but randomly</p>
<p>This code says for each accession in the accession list, sort them randomly (the <code>-R</code> option applied in <code>sort</code>, and then take the first 20 accession numbers from that randomly sorted list. For each of these 20 accession numbers, pull 20,000 reads from the SRA data and put the output read files into the accession folder. Then deactive the conda environment used to pull these SRA data.</p>
<pre><code>for acc in $(cat SraAccList.txt | sort -R | head -n 20); do
    mkdir -p $acc
    fastq-dump --outdir $acc --split-files -N 10001 -X 30000 $acc
done

# deactivate the conda environment now that you are done downloading data
conda deactivate
</code></pre>

<h4 id="installing-haphpipe_2">Installing HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h4 id="activating-haphpipe_2">Activating HAPHPIPE</h4>
<p><strong>insert from students</strong></p>
<h3 id="haphpipe-suite_3">HAPHPIPE suite</h3>
<p><strong>insert from students</strong></p>
<h4 id="hp_reads_3">hp_reads</h4>
<pre><code>    sample_reads
    trim_reads
    join_reads
    ec_reads
</code></pre>
<h4 id="hp_assemble_3">hp_assemble</h4>
<pre><code>    assemble_denovo
    assemble_amplicons
    assemble_scaffold
    align_reads
    call_variants
    vcf_to_consensus
    refine_assembly
    finalize_assembly
</code></pre>
<h4 id="hp_haplotype_3">hp_haplotype</h4>
<pre><code>    predict_haplo
    ph_parser
</code></pre>
<h4 id="hp_annotate_3">hp_annotate</h4>
<pre><code>    pairwise_align
    extract_pairwise
    annotate_from_ref
</code></pre>
<h3 id="example-usage_20">Example usage</h3>
<h4 id="your-virus-with-both-pipelines-document-all-code-and-explanation_2">Your virus with both pipelines. Document all code and explanation.</h4>
<h5 id="starting-directory-structure_2">Starting directory structure</h5>
<p><strong>insert from students</strong></p>
<h5 id="ending-directory-structure_2">Ending directory structure</h5>
<p><strong>insert from students</strong></p>
<h3 id="helpful-sources_2">Helpful sources</h3>
<pre><code>bash scripting
conda
**insert from students**
</code></pre>
<h3 id="faq_2">FAQ</h3>
<p><strong>insert from students</strong></p>
<p></br></p>
<h2 id="-_3">---</h2>
<h1 id="advanced-usage">Advanced Usage</h1>
<pre><code>individual modules + making your own pipelines w/ bash in advanced usage
</code></pre>
<p></br></p>
<h2 id="-_4">---</h2>
<h1 id="file-extension-table">File extension table</h1>
<table>
<thead>
<tr>
<th>Module</th>
<th>Input</th>
<th>Input Format</th>
<th>Output</th>
<th>Output File Names</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>sample_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Subampled FASTQ file(s)</td>
<td><em>sample_1.fastq</em> <br> <em>sample_2.fastq</em></td>
</tr>
<tr>
<td><strong>trim_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Trimmed FASTQ file(s) and <br> summary from Trimmomatic text file</td>
<td><em>trimmed_1.fastq</em> <br> <em>trimmed_2.fastq</em> <br> <em>trimmed_U.fastq</em> <br> <em>trimmomatic_summary.out</em></td>
</tr>
<tr>
<td><strong>join_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>FASTQ files for joined reads and unjoined reads</td>
<td><em>joined.fastq</em> <br> <em>notjoined_1.fastq</em> <br> <em>notjoined_2.fastq</em></td>
</tr>
<tr>
<td><strong>ec_reads</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>Error corrected FASTQ file(s)</td>
<td><em>corrected_1.fastq</em> <br> <em>corrected_2.fastq</em> <br> <em>corrected_U.fastq</em></td>
</tr>
<tr>
<td><strong>assemble_denovo</strong></td>
<td>FASTQ file(s)</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em></td>
<td>De novo assembled contigs FASTA file and <br> de novo summary text file</td>
<td><em>denovo_contigs.fna</em> <br> <em>denovo_summary.txt</em></td>
</tr>
<tr>
<td><strong>assemble_amplicons</strong></td>
<td>FASTA files and <br> GTF file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> and <br> <em>.gtf</em></td>
<td>FASTA file with assembled amplicons</td>
<td><em>amplicon_assembly.fna</em></td>
</tr>
<tr>
<td><strong>assemble_scaffold</strong></td>
<td>FASTA files</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA files with scaffolded and aligned sequences and <br> FASTA file with assembled amplicons</td>
<td><em>scaffold_aligned.fa</em> <br> <em>scaffold_assembly.fa</em> <br> <em>scaffold_imputed.fa</em> <br> <em>scaffold_padded.out</em></td>
</tr>
<tr>
<td><strong>align_reads</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>Aligned BAM file and <br> Bowtie2 alignment output summary text file</td>
<td><em>aligned.bam</em> <br> <em>aligned.bt2.out</em></td>
</tr>
<tr>
<td><strong>call_variants</strong></td>
<td>BAM and <br> FASTA file</td>
<td><em>.bam</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>VCF file with variants</td>
<td><em>variants.vcf.gz</em></td>
</tr>
<tr>
<td><strong>vcf_to_consensus</strong></td>
<td>VCF file</td>
<td><em>.vcf</em></td>
<td>FASTA file with consensus sequence</td>
<td><em>consensus.fna</em></td>
</tr>
<tr>
<td><strong>refine_assembly</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA file with refined consensus sequence</td>
<td><em>refined.fna</em></td>
</tr>
<tr>
<td><strong>finalize_assembly</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>FASTA file with final refined consensus sequence and <br> final BAM file with reads aligned to final FASTA file and <br> VCF file with variants relative to <em>final.fna</em></td>
<td><em>final.fna</em> and <br> <em>final.bam</em> and <br> <em>final.vcf.gz</em></td>
</tr>
<tr>
<td><strong>predict_haplo</strong></td>
<td>FASTQ files and <br> FASTA file</td>
<td><em>.fastq.gz</em> or <br> <em>.fastq</em> or <br> <em>.fq</em> and <br> <em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em></td>
<td>PredictHaplo's fasta-like output file</td>
<td><em>best.fa</em></td>
</tr>
<tr>
<td><strong>ph_parser</strong></td>
<td>PredictHaplo's FAS output file</td>
<td><em>.fas</em></td>
<td>Summary text file with haplotype diversity statistics and <br> FASTA file with haplotype sequences</td>
<td><em>ph_summary.txt</em> and <br> <em>ph_haplotypes.fna</em></td>
</tr>
<tr>
<td><strong>pairwise_align</strong></td>
<td>FASTA files and <br> GTF file</td>
<td><em>.fna</em> or <br> <em>.fasta</em> or <br> <em>.fa</em> and <br> <em>.gtf</em></td>
<td>JSON file</td>
<td><em>pairwise_aligned.json</em></td>
</tr>
<tr>
<td><strong>extract_pairwise</strong></td>
<td>JSON file from pairwise_align output</td>
<td><em>.json</em></td>
<td>FASTA output with region extracted to standard out</td>
<td><em>stdout.fasta</em></td>
</tr>
<tr>
<td><strong>annotate_from_ref</strong></td>
<td>JSON file from pairwise_align output and GTF file</td>
<td><em>.json</em> and <br> <em>.gtf</em></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p></br></p>
<h2 id="-_5">---</h2>
<h1 id="helpful-resources">Helpful resources</h1>
<ul>
<li>Conda <ul>
<li><a href="https://bioconda.github.io">Bioconda</a></li>
<li><a href="https://bioconda.github.io/user/install.html#set-up-channels">Bioconda channels</a></li>
</ul>
</li>
<li>Bash <ul>
<li><a href="https://www.codecademy.com/learn/learn-the-command-line">Command line tutorial</a></li>
<li><a href="https://www.codecademy.com/articles/command-line-commands">List of bash commands</a></li>
<li><a href="https://www.codecademy.com/learn/learn-the-command-line/modules/bash-scripting">Bash scripting tutorial</a></li>
<li>Additional scripting help:  <ul>
<li><a href="https://linuxconfig.org/bash-scripting-tutorial-for-beginners">linuxconfig.org</a></li>
<li><a href="https://flaviocopes.com/bash-scripting/">flaviocopesc.com</a></li>
<li><a href="https://ryanstutorials.net/bash-scripting-tutorial/">ryanstutorials.net</a></li>
</ul>
</li>
</ul>
</li>
<li>NGS overview:<ul>
<li><a href="https://learn.gencore.bio.nyu.edu">NYU resource</a></li>
</ul>
</li>
<li>Accessing on a PC using a virtual machine<ul>
<li><a href="https://blog.storagecraft.com/the-dead-simple-guide-to-installing-a-linux-virtual-machine-on-windows/">Guide to options</a></li>
<li><a href="https://www.virtualbox.org">VirtualBox</a>
</br></li>
</ul>
</li>
</ul>
<h2 id="-_6">---</h2>
<h1 id="faq_3">FAQ</h1>
<ul>
<li>what is directory structure?</li>
</ul>
<p></br></p>
<h2 id="-_7">---</h2>
<h1 id="related-software">Related software</h1>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
