 Module                 | Input                                | Input Format                                          | Output                                                                  | Output File Names                                                                                       |
|------------------------|--------------------------------------|-------------------------------------------------------|-------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|
| **sample_reads**       | FASTQ file(s)                        | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_            | Subampled FASTQ file(s)                                                 | *sample_1.fastq* <br> *sample_2.fastq*                                                                  |
| **trim_reads**         | FASTQ file(s)                        | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_            | Trimmed FASTQ file(s) and <br> summary from Trimmomatic text file       | *trimmed_1.fastq* <br> *trimmed_2.fastq* <br> *trimmed_U.fastq* <br> *trimmomatic_summary.out*          |
| **join_reads**         | FASTQ file(s)                        | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_            | FASTQ files for joined reads and unjoined reads                         | _joined.fastq_ <br> *notjoined_1.fastq* <br> *notjoined_2.fastq*                                        |
| **ec_reads**           | FASTQ file(s)                        | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_            | Error corrected FASTQ file(s)                                           | *corrected_1.fastq* <br> *corrected_2.fastq* <br> *corrected_U.fastq*                                   |
| **assemble_denovo**    | FASTQ file(s)                        | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_            | De novo assembled contigs FASTA file and <br> de novo summary text file | *denovo_contigs.fna* <br> *denovo_summary.txt*                                                          |
| **assemble_amplicons** | FASTA files and <br> GTF file        | _.fna_ or <br> _.fasta_ or <br> _.fa_ and <br> _.gtf_ | FASTA file with assembled amplicons                                     | *amplicon_assembly.fna*                                                                                 |
| **assemble_scaffold**  | FASTA files                          | _.fna_ or <br> _.fasta_ or <br> _.fa_                 | FASTA files with scaffolded and aligned sequences and <br> FASTA file with assembled amplicons | *scaffold_aligned.fa* <br> *scaffold_assembly.fa* <br> *scaffold_imputed.fa* <br> *scaffold_padded.out* |
| **align_reads**        | FASTQ files and <br> FASTA file      | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | Aligned BAM file and <br> Bowtie2 alignment output summary text file | _aligned.bam_ <br> _aligned.bt2.out_                                |
| **call_variants**      | BAM and <br> FASTA file              | _.bam_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | VCF file with variants                                                  | _variants.vcf.gz_                                                                                       |
| **vcf_to_consensus**   | VCF file                             | _.vcf_                                                | FASTA file with consensus sequence                                      | _consensus.fna_                                                                                         |
| **refine_assembly**    | FASTQ files and <br> FASTA file      | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | FASTA file with refined consensus sequence | _refined.fna_                                                       |
| **finalize_assembly**  | FASTQ files and <br> FASTA file      | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | FASTA file with final refined consensus sequence and <br> final BAM file with reads aligned to final FASTA file and <br> VCF file with variants relative to *final.fna* | _final.fna_ and <br> _final.bam_ and <br> _final.vcf.gz_            |
| **predict_haplo**      | FASTQ files and <br> FASTA file      | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | PredictHaplo's fasta-like output file| _best.fa_                                                           |
| **ph_parser**          | PredictHaplo's FAS output file       | _.fas_                                                | Summary text file with haplotype diversity statistics and <br> FASTA file with haplotype sequences | *ph_summary.txt* and <br> *ph_haplotypes.fna*                                                           |
| **clique_snv** | FASTQ files and <br> FASTA file | _.fastq.gz_ or <br> _.fastq_ or <br> _.fq_ and <br> _.fna_ or <br> _.fasta_ or <br> _.fa_ | FASTA file with haplotype sequences and TXT files with summary information | *cs0X_REGION.fasta* <br> *cs0X_REGION.txt* <br> *cs0X_REGION_summary.txt* |
| **pairwise_align**     | FASTA files and <br> GTF file        | _.fna_ or <br> _.fasta_ or <br> _.fa_ and <br> _.gtf_ | JSON file                                                               | *pairwise_aligned.json*                                                                                 |
| **extract_pairwise**   | JSON file from pairwise_align output | *.json*                                               | FASTA output with region extracted to standard out                      | *stdout.fasta*                                                                                          |
| **summary_stats** | Direcoty lists for assembly and haplotype files (TXT) |  _.txt_ | Summary statistics (TXT and TSV) | *summary_stats.txt* <br> *summary_stats.tsv* <br> *PH_summary_stats.tsv*|
| **multiple_align** | FASTA files and/or <br> directory list (TXT) <br> and GTF file | _.fna_ or <br> _.fasta_ or <br> _.fa_ and _.txt_ <br> and <br> _.gtf_ | FASTA alignment and sequence files separated by amplicons in GTF | *alignment_regionX.fasta*  <br> *all_sequences_regionX.fasta* |
|**model_test** | FASTA or PHYLIP alignment file | _.fna_ or <br> _.fasta_ or <br> _.fa_ or <br> _.phy_ | Text file containing best-fit models |_modeltest_results.out_ |
| **build_tree** | FASTA or PHYLIP alignment file | _.fna_ or <br> _.fasta_ or <br> _.fa_ | RAxML output files | *RaxML_info.build_tree.tre* and RAxML tree files |

