<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Modules - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Modules";
    var mkdocs_page_input_path = "modules.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_assemble/">Assemble</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Annotate</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../phylo/">Phylo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../inout/">Input/Output</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demos</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Modules</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="hp_reads">hp_reads</h2>
<p>Stages to manipulate reads. Use -h after any command for a list of options. </p>
<h3 id="sample_reads">sample_reads</h3>
<p>Subsample reads using seqtk (<a href="https://github.com/lh3/seqtk">seqtk documentation</a>).</p>
<p><em>Output files:</em> <br> sample_1.fastq <br> sample_2.fastq</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Sample Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--nreads</td>
<td>Number of reads to sample. If greater than the number of reads in file, all reads will be sampled.</td>
</tr>
<tr>
<td>--frac</td>
<td>Fraction of reads to sample, between 0 and 1. Each read has [frac] probability of being sampled, so the number of sampled reads is not precisely [frac * number of reads].</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr), default is False.</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run, default is False.</td>
</tr>
</tbody>
</table>
<h3 id="trim_reads">trim_reads</h3>
<p>Trim reads using Trimmomatic (<a href="https://github.com/timflutre/trimmomatic">Trimmomatic documentation</a>).</p>
<p><em>Output files:</em> <br> trimmed_1.fastq <br> trimmed_2.fastq <br> trimmed_U.fastq <br> trimmomatic_summary.out</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Trimmomatic Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--adapter_file</td>
<td>Adapter file.</td>
</tr>
<tr>
<td>--trimmers</td>
<td>Trim commands for trimmomatic (default: ['LEADING:3', 'TRAILING:3', 'SLIDINGWINDOW:4:15', 'MINLEN:36']).</td>
</tr>
<tr>
<td>--encoding</td>
<td>Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="join_reads">join_reads</h3>
<p>Join reads using FLASH (<a href="https://github.com/dstreett/FLASH2">FLASH documentation</a>).</p>
<p><em>Output files:</em> <br> joined.fastq <br> notjoined_1.fastq <br> notjoined_2.fastq</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>FLASH Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_overlap</td>
<td>The minimum required overlap length between two reads to provide a confident overlap (default: 10).</td>
</tr>
<tr>
<td>--max_overlap</td>
<td>Maximum overlap length expected in approximately 90% of read pairs, longer overlaps are penalized.</td>
</tr>
<tr>
<td>--allow_outies</td>
<td>Also try combining read pairs in the "outie" orientation (default: False).</td>
</tr>
<tr>
<td>--encoding</td>
<td>Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="ec_reads">ec_reads</h3>
<p>Error correct reads using spades (<a href="https://github.com/ablab/spades">Spades documentation</a>).</p>
<p><em>Output files:</em> <br> corrected_1.fastq <br> corrected_2.fastq <br> corrected_U.fastq</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run, default is False.</td>
</tr>
</tbody>
</table>
<h2 id="hp_assemble">hp_assemble</h2>
<p>Stages to assemble consensus sequence(s). Use -h after any command for a list of options.</p>
<h3 id="assemble_denovo">assemble_denovo</h3>
<p>Assemble reads using denovo assembly.</p>
<p><em>Output files:</em> <br> denovo_contigs.fna <br> denovo_summary.txt</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Assembly Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--no_error_correction</td>
<td>Do not perform error correction (default: False)</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for assembly</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling)</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="assemble_amplicons">assemble_amplicons</h3>
<p>Assemble contigs using reference and amplicon regions.</p>
<p><em>Output files:</em> <br> amplicon_assembly.fna</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Scaffold Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
<tr>
<td>--padding</td>
<td>Bases to include outside reference annotation (default: 50).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="assemble_scaffold">assemble_scaffold</h3>
<p>Scaffold contigs using reference sequence.</p>
<p><em>Output files:</em> <br> scaffold_aligned.fa <br> scaffold_assembly.fa <br> scaffold_imputed.fa <br> scaffold_padded.out</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Scaffold Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqname</td>
<td>Name to append to scaffold sequence (default: sample01).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="align_reads">align_reads</h3>
<p>Align reads to reference.</p>
<p><em>Output files:</em> <br> aligned.bam <br> aligned.bt2.out</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa.</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Alignment Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast, fast, sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local}</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
<tr>
<td>--no_realign</td>
<td>Do not realign indels (default: False).</td>
</tr>
<tr>
<td>--remove_duplicates</td>
<td>Remove duplicates from final alignment. Otherwise duplicates are marked but not removed (default: False).</td>
</tr>
<tr>
<td>--encoding</td>
<td>{Phred+33,Phred+64} Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="call_variants">call_variants</h3>
<p>Call variants.</p>
<p><em>Output files:</em> <br> variants.vcf.gz</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--aln_bam</td>
<td>Alignment file.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Variant Calling Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--emit_all</td>
<td>Output calls for all site (default: False).</td>
</tr>
<tr>
<td>--min_base_qual</td>
<td>Minimum base quality required to consider a base for calling (default: 15).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="vcf_to_consensus">vcf_to_consensus</h3>
<p>Create consensus sequence from VCF.</p>
<p><em>Output files:</em> <br> consensus.fna</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--vcf</td>
<td>VCF file (created with all sites).</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
<tr>
<td>--sampidx</td>
<td>Index for sample if multi-sample VCF (default: 0).</td>
</tr>
</tbody>
</table>
<p><em>Variant Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_DP</td>
<td>Minimum depth to call site (default: 1).</td>
</tr>
<tr>
<td>--major</td>
<td>Allele fraction to make unambiguous call (default: 0.5).</td>
</tr>
<tr>
<td>--minor</td>
<td>Allele fraction to make ambiguous call (default: 0.2).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
</tbody>
</table>
<h3 id="refine_assembly">refine_assembly</h3>
<p>Three step assembly refinement: align reads, call variants, and update reference.</p>
<p><em>Output files:</em> <br> refined.fna</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa.</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Refinement Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--max_step</td>
<td>Maximum number of refinement steps (default: 1).</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for refinement.</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="finalize_assembly">finalize_assembly</h3>
<p>Finalize consensus sequence, align all reads to consensus, and call variants in dataset.</p>
<p><em>Output files:</em> <br> final.fna <br> final.ban <br> final.vcf.gz</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Consensus fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Consensus Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast,fast,sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local} Bowtie2 preset to use (default: very-sensitive).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h2 id="hp_annotate">hp_annotate</h2>
<p>Stages to annotate consensus sequence(s). Use -h after any command for a list of options.</p>
<h3 id="pairwise_align">pairwise_align</h3>
<p>Align amplicons to reference.</p>
<p><em>Output files:</em> <br> 
pairwise_aligned.json</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--amplicons_fa</td>
<td>Fasta file with assembled amplicons.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions. Primary and alternate coding regions should be provided in the attribute field (for amino acid alignment).</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="extract_pairwise">extract_pairwise</h3>
<p>Extract sequence regions from pairwise alignment.</p>
<p><em>Output files:</em> <br> 
stdout.fasta</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--align_json</td>
<td>JSON file describing alignment (output of <em>pairwise_align</em> stage).</td>
</tr>
<tr>
<td>--outfile</td>
<td>Output file (default: stdout).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--outfmt</td>
<td>Format for output: nuc_fa, aln_fa, amp_gtf, ost, or prot_fa (default: nuc_fa).</td>
</tr>
<tr>
<td>--refreg</td>
<td>Reference region. String format is ref:start-stop. For example, the region string to extract pol when aligned to HXB2 is HIV_B.K03455.HXB2:2085-5096.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<h3 id="hp_annotate_from_ref">hp_annotate_from_ref</h3>
<p>Extract sequence regions from pairwise alignment.</p>
<p><em>Output files:</em> <br> </p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--align_json</td>
<td>JSON file describing alignment (output of <em>pairwise_align</em> stage).</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions.</td>
</tr>
<tr>
<td>--outfile</td>
<td>Output file (default: stdout).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
