<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Assemble - HAPHPIPE</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Assemble";
    var mkdocs_page_input_path = "hp_assemble.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> HAPHPIPE</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../install/">Installation</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_reads/">Reads</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Assemble</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#assemble_denovo">assemble_denovo</a></li>
    

    <li class="toctree-l2"><a href="#assemble_amplicons">assemble_amplicons</a></li>
    

    <li class="toctree-l2"><a href="#assemble_scaffold">assemble_scaffold</a></li>
    

    <li class="toctree-l2"><a href="#align_reads">align_reads</a></li>
    

    <li class="toctree-l2"><a href="#call_variants">call_variants</a></li>
    

    <li class="toctree-l2"><a href="#vcf_to_consensus">vcf_to_consensus</a></li>
    

    <li class="toctree-l2"><a href="#refine_assembly">refine_assembly</a></li>
    

    <li class="toctree-l2"><a href="#finalize_assembly">finalize_assembly</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_annotate/">Annotate</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../hp_haplotype/">Haplotype</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../phylo/">Phylo</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../expipes/">Example Pipelines</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../inout/">File Input/Output</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../demos/">Demos</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../adv/">Advanced Users</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../faq/">FAQ</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../help/">Helpful Resources</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">HAPHPIPE</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Assemble</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>Stages in <em>hp_assemble</em> are designed to construct consensus sequence(s). Input reads (in FASTQ format) are assembled 
using either denovo assembly or reference-based alignment. Resulting consensus can be further refined. Use -h after any command for a list of options.</p>
<h3 id="assemble_denovo"><em>assemble_denovo</em></h3>
<p>Assemble reads via de novo assembly using SPAdes (<a href="http://cab.spbu.ru/software/spades/">documentation</a>). Input is reads in FASTQ format. Output is contigs in FNA format.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_denovo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_denovo [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>denovo_contigs.fna</code> <br> <code>denovo_summary.txt</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--no_error_correction</td>
<td>Do not perform error correction (default: False)</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for assembly</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling)</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe assemble_denovo --fq1 corrected_1.fastq --fq2 corrected_2.fastq --outdir denovo_assembly --no_error_correction TRUE
</code></pre>

<h3 id="assemble_amplicons"><em>assemble_amplicons</em></h3>
<p>Assemble contigs from de novo assembly using both a reference sequence and amplicon regions with MUMMER 3+ (<a href="http://mummer.sourceforge.net/manual/">documentation</a>). Input is contigs and reference sequence in FASTA format and amplicon regions in GTF format.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_amplicons [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_amplicons [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; --ref_gtf &lt;GTF&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>amplicon_assembly.fna</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--ref_gtf</td>
<td>GTF format file containing amplicon regions.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Scaffold Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
<tr>
<td>--padding</td>
<td>Bases to include outside reference annotation (default: 50).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe assemble_amplicons --contigs_fa denovo_contigs.fa --ref_fa HIV_B.K03455.HXB2.fasta --ref_gtf HIV_B.K03455.HXB2.gtf
</code></pre>

<h3 id="assemble_scaffold"><em>assemble_scaffold</em></h3>
<p>Scaffold contigs against a reference sequence with MUMMER 3+ (<a href="http://mummer.sourceforge.net/manual/">documentation</a>). Input is contigs in FASTA format and reference sequence in FASTA format. Output is scaffold assembly, alligned scaffold, imputed scaffold, and padded scaffold in FASTA format.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe assemble_scaffold [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_assemble_scaffold [OPTIONS] [SETTINGS] --contigs_fa &lt;FASTA&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <code>scaffold_aligned.fa</code> <br> <code>scaffold_assembly.fa</code> <br> <code>scaffold_imputed.fa</code> <br> <code>scaffold_padded.out</code></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--contigs_fa</td>
<td>Fasta file with assembled contigs.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Fasta file with reference genome to scaffold against.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--seqname</td>
<td>Name to append to scaffold sequence (default: sample01).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe assemble_scaffold --contigs_fa denovo_contigs.fa --ref_fa HIV_B.K03455.HXB2.fasta
</code></pre>

<h3 id="align_reads"><em>align_reads</em></h3>
<p>Map reads to reference sequence (instead of running de novo assembly) using Bowtie2 (<a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">documentation</a>) and Picard (<a href="https://broadinstitute.github.io/picard/">documentation</a>). Input is reads in FASTQ format and reference sequence in FASTA format. </p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe align_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_align_reads [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>aligned.bam</em> <br> <em>aligned.bt2.out</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast, fast, sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local}</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
<tr>
<td>--no_realign</td>
<td>Do not realign indels (default: False).</td>
</tr>
<tr>
<td>--remove_duplicates</td>
<td>Remove duplicates from final alignment. Otherwise duplicates are marked but not removed (default: False).</td>
</tr>
<tr>
<td>--encoding</td>
<td>{Phred+33,Phred+64} Quality score encoding.</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe align_reads --fq1 corrected_1.fastq --fq2 corrected _2.fastq --ref_fa HIV_B.K03455.HXB2.fasta
</code></pre>

<h3 id="call_variants"><em>call_variants</em></h3>
<p>Variant calling from alignment using GATK (<a href="https://software.broadinstitute.org/gatk/download/archive">documentation</a>). Input is alignment file in BAM format and reference sequence in FASTA format (either reference from reference-based assembly or consensus final sequence from de novo assembly). Output is a Variant Call File (VCF) format file. </p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe call_variants [OPTIONS] [SETTINGS] --aln_bam &lt;BAM&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_call_variants [OPTIONS] [SETTINGS] --aln_bam &lt;BAM&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>variants.vcf.gz</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--aln_bam</td>
<td>Alignment file.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--emit_all</td>
<td>Output calls for all site (default: False).</td>
</tr>
<tr>
<td>--min_base_qual</td>
<td>Minimum base quality required to consider a base for calling (default: 15).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe call_variants --aln_bam alignment.bam --ref_fa HIV_B.K03455.HXB2.fasta
</code></pre>

<h3 id="vcf_to_consensus"><em>vcf_to_consensus</em></h3>
<p>Generate a consensus sequence from a VCF file. Input is a VCF file. Output is the consensus sequence in FASTA format. </p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe vcf_to_consensus [OPTIONS] [SETTINGS] --vcf &lt;FASTQ&gt; [--outdir] [--sampidx]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_vcf_to_consensus [OPTIONS] [SETTINGS] --vcf &lt;FASTQ&gt; [--outdir] [--sampidx]</code></p>
<p><em>Output files:</em> <br> <em>consensus.fna</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--vcf</td>
<td>VCF file (created with all sites).</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
<tr>
<td>--sampidx</td>
<td>Index for sample if multi-sample VCF (default: 0).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--min_DP</td>
<td>Minimum depth to call site (default: 1).</td>
</tr>
<tr>
<td>--major</td>
<td>Allele fraction to make unambiguous call (default: 0.5).</td>
</tr>
<tr>
<td>--minor</td>
<td>Allele fraction to make ambiguous call (default: 0.2).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe vcf_to_consensus --vcf variants.vcf
</code></pre>

<h3 id="refine_assembly"><em>refine_assembly</em></h3>
<p>Map reads to a denovo assembly or reference alignment. Assembly or alignment is iteratively updated. Input is reads in FASTQ format and reference sequence (assembly or reference alignment) in FASTA format. Output is refined assembly in FASTA format.</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe refine_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_refine_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> <em>refined.fna</em></p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Reference fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--max_step</td>
<td>Maximum number of refinement steps (default: 1).</td>
</tr>
<tr>
<td>--subsample</td>
<td>Use a subsample of reads for refinement.</td>
</tr>
<tr>
<td>--seed</td>
<td>Seed for random number generator (ignored if not subsampling).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID. Used as read group ID in BAM (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPUs to use (default: 1).</td>
</tr>
<tr>
<td>--xmx</td>
<td>Maximum heap size for Java VM, in GB (default: 32).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Additional options (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe refine_assembly --fq_1 corrected_1.fastq --fq2 corrected_2.fastq --ref_fa HIV_B.K03455.HXB2.fasta
</code></pre>

<h3 id="finalize_assembly"><em>finalize_assembly</em></h3>
<p>Finalize consensus, map reads to consensus, and call variants. Input is reads in FASTQ format and reference sequence in FASTA format. Output is finalized reference sequence, alignment, and variants (in FASTA, BAM, and VCF formats, respectively).</p>
<p><strong>Usage:</strong></p>
<p><code>haphpipe finalize_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><strong>(or):</strong></p>
<p><code>hp_finalize_assembly [OPTIONS] [SETTINGS] --fq1 &lt;FASTQ&gt; --fq2 &lt;FASTQ&gt; --fqU &lt;FASTQ&gt; --ref_fa &lt;FASTA&gt; [--outdir]</code></p>
<p><em>Output files:</em> <br> final.fna <br> final.ban <br> final.vcf.gz</p>
<p><em>Input/Output Arguments:</em> </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--fq1</td>
<td>Fastq file with read 1.</td>
</tr>
<tr>
<td>--fq2</td>
<td>Fastq file with read 2.</td>
</tr>
<tr>
<td>--fqU</td>
<td>Fastq file with unpaired reads.</td>
</tr>
<tr>
<td>--ref_fa</td>
<td>Consensus fasta file.</td>
</tr>
<tr>
<td>--outdir</td>
<td>Output directory (default: current directory).</td>
</tr>
</tbody>
</table>
<p><em>Options:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--bt2_preset</td>
<td>{very-fast,fast,sensitive,very-sensitive,very-fast-local,fast-local,sensitive-local,very-sensitive-local} Bowtie2 preset to use (default: very-sensitive).</td>
</tr>
<tr>
<td>--sample_id</td>
<td>Sample ID (default: sampleXX).</td>
</tr>
</tbody>
</table>
<p><em>Settings:</em></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>--ncpu</td>
<td>Number of CPU to use (default: 1).</td>
</tr>
<tr>
<td>--keep_tmp</td>
<td>Keep temporary directory (default: False).</td>
</tr>
<tr>
<td>--quiet</td>
<td>Do not write output to console (silence stdout and stderr) (default: False).</td>
</tr>
<tr>
<td>--logfile</td>
<td>Append console output to this file.</td>
</tr>
<tr>
<td>--debug</td>
<td>Print commands but do not run (default: False).</td>
</tr>
</tbody>
</table>
<p><em>Example usage:</em></p>
<pre><code>haphpipe finalize_assembly --fq_1 corrected_1.fastq --fq2 corrected_2.fastq --ref_fa refined.fna 
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../hp_annotate/" class="btn btn-neutral float-right" title="Annotate">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../hp_reads/" class="btn btn-neutral" title="Reads"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../hp_reads/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../hp_annotate/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
